---
output: html_document
editor_options: 
  chunk_output_type: console
---
# DArT and dartR background {.unnumbered}

## DArT
![](images/DArT_logo.png)

**Diversity Arrays Technology** (DArT) is a private company that specializes in genotyping-by-sequencing. Their approach is one of genome complexity reduction. 

### Sequencing 

DArTSeq is a method that extracts reproducible genomic variation across the genomes of many individuals at an affordable cost. The technique digests genomic DNA using pairs of restriction enzymes (cutters) (Figure 1). When the DNA is cut at two locations within a reasonable distance of each other, the fragment is available for sequencing using the Illumina short-read platforms. Hence, the data are representational in the sense that they are generated for a random but reproducible selection of small fragments of sequence only, fragments that exhibit variation at the level of single base pairs (SNPs).

### Genotyping

Sequences generated by Diversity Arrays Technology (DArT) are processed using proprietary analytical pipelines before the report (containing the SNP and SilicoDArT markers) is provided to the client and the data are passed to dartR.\

1.  Poor quality sequences are first filtered (minimum barcode Phred score 30, pass percentage 75; minimum whole-read Phred score 10, pass percentage 50). In that way, assignment of the sequences to specific samples in the sample disaggregation step is very reliable.\

2.  These sequences are truncated to 69 bp and aggregated into clusters by the DArT fast clustering algorithm.\

3.  The sequences are then error-corrected. Identical sequences are then collapsed. These error-corrected sequences are analysed using DART software (DArT GS14 pipeline) to output candidate SNP markers.\

4.  As an additional quality control, a selection of samples (30 to 100% of samples) is processed twice. These are technical replicates. Scoring consistency (reproducibility or more strictly, repeatability) is used as the main selection criterion for high quality/low error rate markers.

Where a mutation in present at a site in your genome (across your samples), this is called a **Single Nucleotide Polymorphism** (SNP)

![](images/SNP.jpg)



## dartR

The packages from `dartRverse` pick up the analysis at this point. The additional filtering that you may choose to undertake using dartRverse depends upon the research questions and other considerations as outlined below.

![](images/dartR7.png)

### Data format

dartR used the genlight (and genind) formats from the `adegenet` package @jombart_adegenet.

```{r}
bandicoot.gl <- dartR.data::bandicoot.gl
```


  1. Basically, this object is a dataframe which has the individuals in columns and the SNPs in rows

  2. In the image below the SNPs are coded according that nucleotide mutations (e.g. A/A for homozygous, or C/G for heterozygous)
  
![](images/SNP_data_ATGC.png)

  3. But to  make all genetic markers comparable, they will be code:
  
    0 = homozygous for the reference allele
    
    1 = heterozygous
    
    2 = homozygous for the alternate allele
    
    NA = missing data

![](images/SNP_data_012.png)


  4. The genlight object also stored the metadata of your individuals and SNPs
  
:::{.panel-tabset}

#### Individual metadata

```{r}
bandicoot.gl@pop
bandicoot.gl@ind.names
knitr::kable(head(bandicoot.gl@other$ind.metrics))
```

#### SNP metadata

```{r}
bandicoot.gl@n.loc
knitr::kable(head(bandicoot.gl@other$loc.metrics))
```

:::