<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>OCS2024 - Exploring the genome - an overview of tools and applications for the modern Chondrichthyes ecologist - Let’s get started</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./references.html" rel="next">
<link href="./DArT_dartR.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="css.css">
<link rel="stylesheet" href="include/webex.css">
</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./main_code.html">Let’s get started</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">OCS2024 - Exploring the genome - an overview of tools and applications for the modern Chondrichthyes ecologist</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome!</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./install.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Install packages</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./refreshers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Refreshers and dartR tutorials</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./DArT_dartR.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">DArT and dartR background</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./main_code.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Let’s get started</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#load-packages" id="toc-load-packages" class="nav-link active" data-scroll-target="#load-packages">Load packages</a></li>
  <li><a href="#data-introduction" id="toc-data-introduction" class="nav-link" data-scroll-target="#data-introduction">Data introduction</a>
  <ul class="collapse">
  <li><a href="#bull-shark" id="toc-bull-shark" class="nav-link" data-scroll-target="#bull-shark">Bull shark</a></li>
  <li><a href="#northern-river-shark" id="toc-northern-river-shark" class="nav-link" data-scroll-target="#northern-river-shark">Northern River shark</a></li>
  <li><a href="#blue-shark" id="toc-blue-shark" class="nav-link" data-scroll-target="#blue-shark">Blue shark</a></li>
  </ul></li>
  <li><a href="#reading-and-filtering-raw-data" id="toc-reading-and-filtering-raw-data" class="nav-link" data-scroll-target="#reading-and-filtering-raw-data">Reading and filtering raw data</a>
  <ul class="collapse">
  <li><a href="#import-data" id="toc-import-data" class="nav-link" data-scroll-target="#import-data">Import data</a></li>
  <li><a href="#save-raw-data" id="toc-save-raw-data" class="nav-link" data-scroll-target="#save-raw-data">Save raw data</a></li>
  <li><a href="#quality-filtering" id="toc-quality-filtering" class="nav-link" data-scroll-target="#quality-filtering">Quality filtering</a>
  <ul class="collapse">
  <li><a href="#average-reproducibility" id="toc-average-reproducibility" class="nav-link" data-scroll-target="#average-reproducibility">Average reproducibility</a></li>
  <li><a href="#invariable-sites" id="toc-invariable-sites" class="nav-link" data-scroll-target="#invariable-sites">Invariable sites</a></li>
  <li><a href="#read-depth" id="toc-read-depth" class="nav-link" data-scroll-target="#read-depth">Read depth</a></li>
  <li><a href="#missing-data-for-snps" id="toc-missing-data-for-snps" class="nav-link" data-scroll-target="#missing-data-for-snps">Missing data for SNPs</a></li>
  <li><a href="#missing-data-for-individuals" id="toc-missing-data-for-individuals" class="nav-link" data-scroll-target="#missing-data-for-individuals">Missing data for individuals</a></li>
  <li><a href="#individual-heterozygosity" id="toc-individual-heterozygosity" class="nav-link" data-scroll-target="#individual-heterozygosity">Individual heterozygosity</a></li>
  <li><a href="#invariable-sites-1" id="toc-invariable-sites-1" class="nav-link" data-scroll-target="#invariable-sites-1">Invariable sites</a></li>
  <li><a href="#linked-loci" id="toc-linked-loci" class="nav-link" data-scroll-target="#linked-loci">Linked loci</a></li>
  <li><a href="#hardy-weinberg-equilibrium" id="toc-hardy-weinberg-equilibrium" class="nav-link" data-scroll-target="#hardy-weinberg-equilibrium">Hardy-Weinberg Equilibrium</a></li>
  <li><a href="#remove-duplicate-individuals" id="toc-remove-duplicate-individuals" class="nav-link" data-scroll-target="#remove-duplicate-individuals">Remove duplicate individuals</a></li>
  </ul></li>
  <li><a href="#filtering-conclusion" id="toc-filtering-conclusion" class="nav-link" data-scroll-target="#filtering-conclusion">Filtering Conclusion</a></li>
  </ul></li>
  <li><a href="#population-structure" id="toc-population-structure" class="nav-link" data-scroll-target="#population-structure">Population structure</a>
  <ul class="collapse">
  <li><a href="#principal-component-analysis" id="toc-principal-component-analysis" class="nav-link" data-scroll-target="#principal-component-analysis">Principal Component Analysis</a></li>
  <li><a href="#fixation-indices" id="toc-fixation-indices" class="nav-link" data-scroll-target="#fixation-indices">Fixation indices</a></li>
  </ul></li>
  <li><a href="#genetic-diversity" id="toc-genetic-diversity" class="nav-link" data-scroll-target="#genetic-diversity">Genetic diversity</a></li>
  <li><a href="#population-connectivity" id="toc-population-connectivity" class="nav-link" data-scroll-target="#population-connectivity">Population connectivity</a></li>
  <li><a href="#natural-selection-in-the-context-of-adaptive-divergence" id="toc-natural-selection-in-the-context-of-adaptive-divergence" class="nav-link" data-scroll-target="#natural-selection-in-the-context-of-adaptive-divergence">Natural selection in the context of adaptive divergence</a>
  <ul class="collapse">
  <li><a href="#fst-based-outliers" id="toc-fst-based-outliers" class="nav-link" data-scroll-target="#fst-based-outliers"><span class="math inline">\(FST\)</span>-based outliers</a></li>
  <li><a href="#genotype-environment-associations" id="toc-genotype-environment-associations" class="nav-link" data-scroll-target="#genotype-environment-associations">Genotype-Environment associations</a></li>
  </ul></li>
  <li><a href="#relatedness" id="toc-relatedness" class="nav-link" data-scroll-target="#relatedness">Relatedness</a>
  <ul class="collapse">
  <li><a href="#convert-data" id="toc-convert-data" class="nav-link" data-scroll-target="#convert-data">Convert data</a></li>
  <li><a href="#run-coancestry-on-full-data" id="toc-run-coancestry-on-full-data" class="nav-link" data-scroll-target="#run-coancestry-on-full-data">Run coancestry on full data</a></li>
  <li><a href="#one-population" id="toc-one-population" class="nav-link" data-scroll-target="#one-population">One population</a></li>
  </ul></li>
  <li><a href="#sex-linked-markers" id="toc-sex-linked-markers" class="nav-link" data-scroll-target="#sex-linked-markers">Sex-linked markers</a>
  <ul class="collapse">
  <li><a href="#load-data" id="toc-load-data" class="nav-link" data-scroll-target="#load-data">Load data</a>
  <ul class="collapse">
  <li><a href="#bull-shark-1" id="toc-bull-shark-1" class="nav-link" data-scroll-target="#bull-shark-1">Bull Shark</a></li>
  <li><a href="#blue-shark-1" id="toc-blue-shark-1" class="nav-link" data-scroll-target="#blue-shark-1">Blue Shark</a></li>
  <li><a href="#run-gl.filter.sexlinked" id="toc-run-gl.filter.sexlinked" class="nav-link" data-scroll-target="#run-gl.filter.sexlinked">Run <code>gl.filter.sexlinked</code></a></li>
  <li><a href="#run-gl.infer.sex" id="toc-run-gl.infer.sex" class="nav-link" data-scroll-target="#run-gl.infer.sex">Run <code>gl.infer.sex</code></a></li>
  </ul></li>
  </ul></li>
  <li><a href="#session-information" id="toc-session-information" class="nav-link" data-scroll-target="#session-information">Session information</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Let’s get started</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="load-packages" class="level1">
<h1>Load packages</h1>
<p>These are the main packages we will be using today. Please check the <code>Install packages</code> tab for details.</p>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-1_f875840550703051d05ab44e4d4e2ea9">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dartRverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(adegenet)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(diveRsity)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(snpStats) <span class="co">#only for LD</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtern) <span class="co"># for HWE diagnostics</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(related)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data-introduction" class="level1">
<h1>Data introduction</h1>
<p>We have made three dataset available for this tutorial. <strong>Pick one:</strong></p>
<ol type="1">
<li><p>Bull shark DArTcap data</p></li>
<li><p>Northern River shark DArTcap data</p></li>
<li><p>Blue Shark DArTseq data</p></li>
</ol>
<section id="bull-shark" class="level2">
<h2 class="anchored" data-anchor-id="bull-shark">Bull shark</h2>
<p>Data from <span class="citation" data-cites="devloo_bull">Devloo-Delva, Burridge, et al. (<a href="references.html#ref-devloo_bull" role="doc-biblioref">2023</a>)</span>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Bull_Shark.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Bull Shark</figcaption>
</figure>
</div>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-2_d40d22b1bb37043ac35245ee915793c3">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>data.file <span class="ot">&lt;-</span> <span class="st">"data/Bull_shark_DArTcap_1rowGT.csv"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>metadata.file <span class="ot">&lt;-</span> <span class="st">"data/Bull_shark_DArTcap_metadata.csv"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="northern-river-shark" class="level2">
<h2 class="anchored" data-anchor-id="northern-river-shark">Northern River shark</h2>
<p>Data from <span class="citation" data-cites="feutry_glyphis">Feutry et al. (<a href="references.html#ref-feutry_glyphis" role="doc-biblioref">2020</a>)</span>.</p>
<p>Around 470 <em>G. garricki</em> samples were collected from 11 different sampling locations in Northern Australia and Papua New Guinea.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Northern_River_Shark.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Northern River shark</figcaption>
</figure>
</div>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-3_551ed22f717044d64ff8fc9182b2b18d">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>data.file <span class="ot">&lt;-</span> <span class="st">"data/Northern_river_shark_DArTcap_1rowGT.csv"</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>metadata.file <span class="ot">&lt;-</span> <span class="st">"data/Northern_river_shark_DArTcap_metadata.csv"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="blue-shark" class="level2">
<h2 class="anchored" data-anchor-id="blue-shark">Blue shark</h2>
<p>Data from <span class="citation" data-cites="nikolic_blue">Nikolic et al. (<a href="references.html#ref-nikolic_blue" role="doc-biblioref">2023</a>)</span>.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/Blue_shark.jpg" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">The Blue Shark</figcaption>
</figure>
</div>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-4_0da512732589c76b3728cb107fea1eed">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>data.file <span class="ot">&lt;-</span> <span class="st">"data/Blue_shark_DArTSeq_1rowGT_subset.csv"</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>metadata.file <span class="ot">&lt;-</span> <span class="st">"data/Blue_shark_DArTseq_metadata.csv"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="reading-and-filtering-raw-data" class="level1">
<h1>Reading and filtering raw data</h1>
<p>For the purpose of this workshop we will use the <code>dartRverse</code> packages to read and filter the data <span class="citation" data-cites="gruber_dartr mijangos_dartr2">(<a href="references.html#ref-gruber_dartr" role="doc-biblioref">Gruber et al. 2018</a>; <a href="references.html#ref-mijangos_dartr2" role="doc-biblioref">Mijangos et al. 2022</a>)</span>.</p>
<section id="import-data" class="level2">
<h2 class="anchored" data-anchor-id="import-data">Import data</h2>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-5_a8ba1db3a6018683b4ce9ded05f2b4f2">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.read.dart</span>(<span class="at">filename =</span> data.file,</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>                                <span class="at">ind.metafile =</span> metadata.file, </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>                                <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.read.dart 
Starting utils.read.dart 
Completed: utils.read.dart 
Starting utils.dart2genlight 
Ids for individual metadata does not match the number of ids in the SNP data file. Maybe this is fine if a subset matches.
ind.metafile ids not matched were:
 [1] "GGA004"  "GGA025"  "GGA072"  "GGA192"  "GGA223"  "GGA227"  "GGA273" 
 [8] "GGA274"  "GGA275"  "GGA276"  "GGA277"  "GGA278"  "GGA279"  "GGA280" 
[15] "GGA281"  "GGA282"  "GGA283"  "GGA284"  "GGA285"  "GGA286"  "GGA287" 
[22] "GGA288"  "GGA289"  "GGA290"  "GGA291"  "GGA292"  "GGA293"  "GGA294" 
[29] "GGA295"  "GGA296"  "GGA297"  "GGA298"  "GGA299"  "GGA300"  "GGA301" 
[36] "GGA302"  "GGA303"  "GGA312"  "GGA313"  "GGA314"  "GGA315"  "GGA316" 
[43] "GGA317"  "GGA318"  "GGA319"  "GGA320"  "GGA321"  "GGA322"  "GGA323" 
[50] "GGA324"  "GGA325"  "GGA326"  "GGL034"  "GGL035"  "GGL036"  "GGL037" 
[57] "GGL038"  "GGL039"  "GGL040"  "GGL041"  "GGL042"  "GGL043"  "GGL044" 
[64] "GGL045"  "GGL046"  "GGL047"  "GGL048"  "GGL049"  "GGSA012"
Completed: utils.dart2genlight 
Starting gl.compliance.check 
  Processing genlight object with SNP data
  Checking coding of SNPs
    SNP data scored NA, 0, 1 or 2 confirmed
  Checking for population assignments
    Population assignments confirmed
  Checking locus metrics and flags
  Recalculating locus metrics
  Checking for monomorphic loci
    Dataset contains monomorphic loci
  Checking for loci with all missing data
    No loci with all missing data detected
  Checking whether individual names are unique.
  Checking for individual metrics
    Individual metrics confirmed
  Spelling of coordinates checked and changed if necessary to 
            lat/lon
Completed: gl.compliance.check 
Completed: :: 
 Completed: dartR.base 
 Completed: gl.read.dart </code></pre>
</div>
</div>
</section>
<section id="save-raw-data" class="level2">
<h2 class="anchored" data-anchor-id="save-raw-data">Save raw data</h2>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-6_7c745e2cf98b02034b48905748f1f523">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(data.gl, <span class="at">file =</span> <span class="st">"data/raw_genlight.Rdata"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="quality-filtering" class="level2">
<h2 class="anchored" data-anchor-id="quality-filtering">Quality filtering</h2>
<p>The order of filtering can be important and requires some thought. The filtering steps may change depending on the analysis you intend to run. There is no one-fits-all approach to SNP filtering.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Be careful not to over-filter. The objective is to get an appropriate balance between signal to noise ratio, not to eliminate noise altogether at the expense of also taking out some signal. This balance will depend on downstream application.</p>
<p><em>A good strategy is to undertake an exploratory analysis first, whereby you experiment with filtering. Perhaps start with very stringent filtering, examine the impact on the final analysis, then progressively reduce the stringency to select the minimal filtering regime that still delivers a stable outcome. Play with your data to get a feel for the balance between signal to noise ratio, in the context of the analyses you propose to subsequently undertake.</em></p>
</div>
</div>
<section id="average-reproducibility" class="level3">
<h3 class="anchored" data-anchor-id="average-reproducibility">Average reproducibility</h3>
<p>DArT includes several technical replicates (~30 %) of your samples to see how reproducible the results are. Some SNPs are less reliable than others.</p>
<p><em>This metric is based on all individuals in your original dataset.</em></p>
<p><strong>It is not advised to filter on reproducibility &gt; 0.99, since you are like to remove more heterozygous SNPs (i.e.&nbsp;more difficult to genotype consistently).</strong></p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true">Report reproducibility</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">Filter reproducibility</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-7_3a24dc48949443cc8cb1e102c251bdd7">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.reproducibility</span>(data.gl, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.reproducibility 
  Reporting Repeatability by Locus
  No. of loci = 9111 
  No. of individuals = 467 
    Minimum      :  0.788618 
    1st quartile :  0.996667 
    Median       :  1 
    Mean         :  0.9932215 
    3r quartile  :  1 
    Maximum      :  1 
    Missing Rate Overall:  0.07 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Quantile Threshold Retained Percent Filtered Percent
1      100%  1.000000     5944    65.2     3167    34.8
2       95%  1.000000     5944    65.2     3167    34.8
3       90%  1.000000     5944    65.2     3167    34.8
4       85%  1.000000     5944    65.2     3167    34.8
5       80%  1.000000     5944    65.2     3167    34.8
6       75%  1.000000     5944    65.2     3167    34.8
7       70%  1.000000     5944    65.2     3167    34.8
8       65%  1.000000     5944    65.2     3167    34.8
9       60%  1.000000     5944    65.2     3167    34.8
10      55%  1.000000     5944    65.2     3167    34.8
11      50%  1.000000     5944    65.2     3167    34.8
12      45%  1.000000     5944    65.2     3167    34.8
13      40%  1.000000     5944    65.2     3167    34.8
14      35%  1.000000     5944    65.2     3167    34.8
15      30%  0.996753     6579    72.2     2532    27.8
16      25%  0.996667     6846    75.1     2265    24.9
17      20%  0.993506     7492    82.2     1619    17.8
18      15%  0.990909     7746    85.0     1365    15.0
19      10%  0.984127     8200    90.0      911    10.0
20       5%  0.957792     8667    95.1      444     4.9
21       0%  0.788618     9111   100.0        0     0.0
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.reproducibility </code></pre>
</div>
</div>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-8_68b82260bb55919bb628222c51de3a85">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.reproducibility</span>(data.gl, <span class="at">threshold =</span> <span class="fl">0.98</span>, </span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">plot.colors =</span> <span class="fu">c</span>(<span class="st">"#2171B5"</span>, <span class="st">"#6BAED6"</span>),</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>                                                 <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.filter.reproducibility </code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.filter.reproducibility </code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="invariable-sites" class="level3">
<h3 class="anchored" data-anchor-id="invariable-sites">Invariable sites</h3>
<p>You want to remove SNP that are not informative for population-level analyses. This will improve computation time and reduce the noise in the data.</p>
<section id="all-na" class="level4">
<h4 class="anchored" data-anchor-id="all-na">All NA</h4>
<p>Removes SNPs that have missing data for all samples.</p>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-9_92c615e39d5d4a406b291cb1bab2721b">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.allna</span>(data.gl, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.filter.allna 
Completed: :: 
 Completed: dartR.base 
 Completed: gl.filter.allna </code></pre>
</div>
</div>
</section>
<section id="monomorphic" class="level4">
<h4 class="anchored" data-anchor-id="monomorphic">Monomorphic</h4>
<p>Looks for loci which are all 0 (or 1 or 2) over the all samples. These are not informative.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-2-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-1" role="tab" aria-controls="tabset-2-1" aria-selected="true">Report monomorphs</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-2-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-2-2" role="tab" aria-controls="tabset-2-2" aria-selected="false">Filter monomorphs</a></li></ul>
<div class="tab-content">
<div id="tabset-2-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-2-1-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-10_8599987904d78ef1e4553c05b71bc07b">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.monomorphs</span>(data.gl, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.monomorphs 

  No. of loci: 8364 
    Polymorphic loci: 5318 
    Monomorphic loci: 3046 
    Loci scored all NA: 1 
  No. of individuals: 467 
  No. of populations: 5 

Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.monomorphs </code></pre>
</div>
</div>
</div>
<div id="tabset-2-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-2-2-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-11_7b3dd6c458cedc992069c8ab29af97ee">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.monomorphs</span>(data.gl, <span class="at">verbose =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
[dartR.base vers. 0.88 Build = v.2023.2 ]
 Starting dartR.base 
[dartR.base vers. 0.88 Build = v.2023.2 ]
 Starting gl.filter.monomorphs 
[dartR.base vers. 0.88 Build = v.2023.2 ]
  Processing genlight object with SNP data
  Identifying monomorphic loci
  Removing monomorphic loci and loci with all missing 
                       data
    Original No. of loci: 8364 
    Monomorphic loci: 3046 
    Loci scored all NA: 1 
    No. of loci deleted: 3046 
    No. of loci retained: 5318 
    No. of individuals: 467 
    No. of populations: 5 
Completed: :: 
 Completed: dartR.base 
 Completed: gl.filter.monomorphs </code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="minor-allele-frequency" class="level4">
<h4 class="anchored" data-anchor-id="minor-allele-frequency">Minor Allele Frequency</h4>
<p>Some mutations only occur in a couple of individuals, these are not informative for population-level analyses. Loci with a low Minor Allele Frequency (MAF) might not be informative enough and just increase computation time.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>The threshold should be lower than the proportion of your smallest sampling location. For example, if one of your sample sites has 5 samples out of a total of 100 samples, the MAF should not be &lt; 0.05. Else, your risk loosing the informative markers that differntiate that population.</p>
</div>
</div>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-3-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-1" role="tab" aria-controls="tabset-3-1" aria-selected="true">Report maf</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-3-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-3-2" role="tab" aria-controls="tabset-3-2" aria-selected="false">Filter maf</a></li></ul>
<div class="tab-content">
<div id="tabset-3-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-3-1-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-12_a513bda73c6ef959c76fd14997f5d8bc">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.maf</span>(data.gl, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.maf 
Starting :: 

 Starting dartR.base 

 Starting gl.report.maf 

  Reporting Minor Allele Frequency (MAF) by Locus for population Cambridge Gulf 
  No. of loci = 2510 
  No. of individuals = 30 
    Minimum      :  0.0167 
    1st quantile :  0.0345 
    Median       :  0.1667 
    Mean         :  0.1925474 
    3r quantile  :  0.3167 
    Maximum      :  0.5 
    Missing Rate Overall:  0.03 

  Reporting Minor Allele Frequency (MAF) by Locus for population Daly River 
  No. of loci = 2542 
  No. of individuals = 30 
    Minimum      :  0.0167 
    1st quantile :  0.0345 
    Median       :  0.1511 
    Mean         :  0.1870297 
    3r quantile  :  0.3167 
    Maximum      :  0.5 
    Missing Rate Overall:  0.04 

  Reporting Minor Allele Frequency (MAF) by Locus for population King Sound 
  No. of loci = 1516 
  No. of individuals = 19 
    Minimum      :  0.0263 
    1st quantile :  0.0278 
    Median       :  0.1316 
    Mean         :  0.1872923 
    3r quantile  :  0.3158 
    Maximum      :  0.5 
    Missing Rate Overall:  0.03 

  Reporting Minor Allele Frequency (MAF) by Locus for population Papua New Guinea 
  No. of loci = 1642 
  No. of individuals = 4 
    Minimum      :  0.125 
    1st quantile :  0.125 
    Median       :  0.25 
    Mean         :  0.2812373 
    3r quantile  :  0.375 
    Maximum      :  0.5 
    Missing Rate Overall:  0.02 

  Reporting Minor Allele Frequency (MAF) by Locus for population Van Diemen Gulf 
  No. of loci = 4965 
  No. of individuals = 384 
    Minimum      :  0.0013 
    1st quantile :  0.0026 
    Median       :  0.0084 
    Mean         :  0.09757069 
    3r quantile  :  0.1706 
    Maximum      :  0.5 
    Missing Rate Overall:  0.05 

  Reporting Minor Allele Frequency (MAF) by Locus OVERALL
  No. of loci = 5318 
  No. of individuals = 467 
    Minimum      :  0.0011 
    1st quantile :  0.0022 
    Median       :  0.00755 
    Mean         :  0.09431826 
    3r quantile  :  0.15755 
    Maximum      :  0.4989 
    Missing Rate Overall:  0.06 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Quantile Threshold Retained Percent Filtered Percent
1      100%    0.4989        2     0.0     5316   100.0
2       95%    0.4197      267     5.0     5051    95.0
3       90%    0.3462      532    10.0     4786    90.0
4       85%    0.2842      799    15.0     4519    85.0
5       80%    0.2192     1064    20.0     4254    80.0
6       75%    0.1576     1330    25.0     3988    75.0
7       70%    0.1039     1597    30.0     3721    70.0
8       65%    0.0511     1862    35.0     3456    65.0
9       60%    0.0184     2131    40.1     3187    59.9
10      55%    0.0109     2397    45.1     2921    54.9
11      50%    0.0075     2678    50.4     2640    49.6
12      45%    0.0054     2954    55.5     2364    44.5
13      40%    0.0043     3222    60.6     2096    39.4
14      35%    0.0033     3487    65.6     1831    34.4
15      30%    0.0026     3731    70.2     1587    29.8
16      25%    0.0022     4118    77.4     1200    22.6
17      20%    0.0021     4257    80.0     1061    20.0
18      15%    0.0012     4555    85.7      763    14.3
19      10%    0.0011     5030    94.6      288     5.4
20       5%    0.0011     5318   100.0        0     0.0
21       0%    0.0011     5318   100.0        0     0.0
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.maf </code></pre>
</div>
</div>
</div>
<div id="tabset-3-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-3-2-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-13_90a89897b1478e549293fc6c7425e12e">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.maf</span>(data.gl, <span class="at">threshold =</span> <span class="fl">0.005</span>, </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">plot.colors =</span> <span class="fu">c</span>(<span class="st">"#2171B5"</span>, <span class="st">"#6BAED6"</span>),</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.filter.maf </code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.filter.maf </code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="read-depth" class="level3">
<h3 class="anchored" data-anchor-id="read-depth">Read depth</h3>
<p>Read depth has a considerable influence on the veracity of SNP calls. The low values arise because of chance variation in the coverage of particular bases. The high values arise because of the differential efficiency of the PCR steps or paralogous fragments. Under some circumstances, it might be sensible to push the lower threshold for read depth higher than has been used by the Diversity Arrays Technology in their report. Analyses that rely heavily on the accuracy of calls of heterozygotes may require a higher threshold for read depth, say 10x, for example.</p>
<p><strong>Check the histogram to see how many SNP have low or high counts. Low counts might not be reliably genotyped. High counts may be due to paralogous fragments.</strong></p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-4-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-1" role="tab" aria-controls="tabset-4-1" aria-selected="true">Report rdepth</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-4-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-4-2" role="tab" aria-controls="tabset-4-2" aria-selected="false">Filter rdepth</a></li></ul>
<div class="tab-content">
<div id="tabset-4-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-4-1-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-14_d936a17a2ef07db906309d2f02af8751">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.rdepth</span>(data.gl, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.rdepth 
  Reporting Read Depth by Locus
  No. of loci = 3027 
  No. of individuals = 467 
    Minimum      :  2 
    1st quartile :  12.8 
    Median       :  43.2 
    Mean         :  68.76938 
    3r quartile  :  89.9 
    Maximum      :  573.8 
    Missing Rate Overall:  0.05 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   Quantile Threshold Retained Percent Filtered Percent
1      100%     573.8        1     0.0     3026   100.0
2       95%     249.7      152     5.0     2875    95.0
3       90%     156.7      303    10.0     2724    90.0
4       85%     122.9      455    15.0     2572    85.0
5       80%     103.6      606    20.0     2421    80.0
6       75%      90.0      757    25.0     2270    75.0
7       70%      79.3      910    30.1     2117    69.9
8       65%      70.4     1061    35.1     1966    64.9
9       60%      61.6     1211    40.0     1816    60.0
10      55%      51.4     1363    45.0     1664    55.0
11      50%      43.2     1514    50.0     1513    50.0
12      45%      35.8     1666    55.0     1361    45.0
13      40%      28.6     1818    60.1     1209    39.9
14      35%      21.9     1969    65.0     1058    35.0
15      30%      17.1     2120    70.0      907    30.0
16      25%      12.8     2272    75.1      755    24.9
17      20%       9.4     2422    80.0      605    20.0
18      15%       6.9     2576    85.1      451    14.9
19      10%       5.0     2728    90.1      299     9.9
20       5%       3.3     2879    95.1      148     4.9
21       0%       2.0     3027   100.0        0     0.0
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.rdepth </code></pre>
</div>
</div>
</div>
<div id="tabset-4-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-4-2-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-15_80dddb0bc3f7302199618f3f8b1ec03a">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.rdepth</span>(data.gl, <span class="at">lower =</span> <span class="dv">5</span>, <span class="at">upper =</span> <span class="dv">200</span>, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.filter.rdepth </code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.filter.rdepth </code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="missing-data-for-snps" class="level3">
<h3 class="anchored" data-anchor-id="missing-data-for-snps">Missing data for SNPs</h3>
<p>Call rate is defined as <code>1 - proportion of missing data</code>. SNPs with too much missing data are not reliable and informative.</p>
<p>Call Rate for SNPs can arise from two sources. The first source is where a missing value arises because the sequence tag bearing the target SNP cannot be amplified – there has been a mutation at one or both of the restriction sites. The second source of missing values is where the read depth is insufficient to make a reliable call on the SNP. Either way, the SNP is not called and is recorded as NA. A first step in filtering on Call Rate is to examine the distribution of Call Rates across loci. We use the function <code>gl.report.callrate</code> to yield the following output and graph.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-5-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-1" role="tab" aria-controls="tabset-5-1" aria-selected="true">Report SNP callrate</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-5-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-5-2" role="tab" aria-controls="tabset-5-2" aria-selected="false">Filter SNP callrate</a></li></ul>
<div class="tab-content">
<div id="tabset-5-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-5-1-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-16_3eac8fdf80f5657d0bb40fe3549324fe">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.callrate</span>(data.gl, <span class="at">method =</span> <span class="st">"loc"</span>, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.callrate 
  Reporting Call Rate by Locus
  No. of loci = 2513 
  No. of individuals = 467 
    Minimum      :  0.20985 
    1st quartile :  0.987152 
    Median       :  0.993576 
    Mean         :  0.9637389 
    3r quartile  :  1 
    Maximum      :  1 
    Missing Rate Overall:  0.0363 </code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.callrate </code></pre>
</div>
</div>
</div>
<div id="tabset-5-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-5-2-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-17_c3fd3bcf65af1745b95ee0444317155e">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.callrate</span>(data.gl, <span class="at">method =</span> <span class="st">"loc"</span>,  <span class="at">threshold =</span> <span class="fl">0.8</span>, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.filter.callrate </code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.filter.callrate </code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="missing-data-for-individuals" class="level3">
<h3 class="anchored" data-anchor-id="missing-data-for-individuals">Missing data for individuals</h3>
<p>Call rate is defined as <code>1 - proportion of missing data</code>. Samples with degraded DNA will have a lot of missing sequencing data. These samples are unreliably sequenced and should be removed.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-6-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-1" role="tab" aria-controls="tabset-6-1" aria-selected="true">Report ind. callrate</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-6-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-6-2" role="tab" aria-controls="tabset-6-2" aria-selected="false">Filter ind. callrate</a></li></ul>
<div class="tab-content">
<div id="tabset-6-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-6-1-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-18_32fddefeda368851580ef78696cadf85">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.callrate</span>(data.gl, <span class="at">method =</span> <span class="st">"ind"</span>, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.callrate 

  Reporting Call Rate by Individual
  No. of loci = 2373 
  No. of individuals = 467 
    Minimum      :  0.4551201 
    1st quartile :  0.9903076 
    Median       :  0.9928361 
    Mean         :  0.9871529 
    3r quartile  :  0.9945217 
    Maximum      :  0.997893 
    Missing Rate Overall:  0.0128 

Listing 5 populations and their average CallRates
  Monitor again after filtering
        Population CallRate   N
1   Cambridge Gulf   0.9898  30
2       Daly River   0.9829  30
3       King Sound   0.9830  19
4 Papua New Guinea   0.9854   4
5  Van Diemen Gulf   0.9875 384

Listing 20 individuals with the lowest CallRates
  Use this list to see which individuals will be lost on filtering by individual
  Set ind.to.list parameter to see more individuals
   Individual  CallRate
1      GGL032 0.4551201
2     GGSA011 0.4938896
3     GGSA016 0.6392752
4      GGA051 0.6814159
5      GGA053 0.7568479
6      GGD029 0.7926675
7      GGA252 0.9447956
8      GGA241 0.9481669
9      GGA188 0.9603877
10     GGK013 0.9713443
11     GGA176 0.9717657
12     GGL001 0.9721871
13     GGS029 0.9759798
14    GGKS002 0.9768226
15     GGA350 0.9785082
16     GGD011 0.9793510
17     GGK008 0.9797724
18    GGKS005 0.9801938
19       GG19 0.9806153
20     GGK006 0.9814581

)</code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.callrate </code></pre>
</div>
</div>
</div>
<div id="tabset-6-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-6-2-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-19_79a8000deba2d2b3d1d67cb039dc3ae0">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.callrate</span>(data.gl, <span class="at">method =</span> <span class="st">"ind"</span>, <span class="at">threshold =</span> <span class="fl">0.8</span>, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.filter.callrate </code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.filter.callrate </code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="individual-heterozygosity" class="level3">
<h3 class="anchored" data-anchor-id="individual-heterozygosity">Individual heterozygosity</h3>
<p>Delete individuals with a heterozygosity below or above the threshold. High heterozygosity could be due to DNA cross-contamination from other individuals. Low heterozygosity could be due to population structure, inbreeding or excess missing values (normally the latter should be filtered out already).</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-7-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-1" role="tab" aria-controls="tabset-7-1" aria-selected="true">Report ind. heterozygosity</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-7-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-7-2" role="tab" aria-controls="tabset-7-2" aria-selected="false">Filter ind. heterozygosity</a></li></ul>
<div class="tab-content">
<div id="tabset-7-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-7-1-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-20_65559f63fdac737bc35eb11ab89e872d">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.heterozygosity</span>(data.gl, <span class="at">method =</span> <span class="st">"ind"</span>, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.heterozygosity 
Starting gl.colors 
Selected color type 2 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.heterozygosity </code></pre>
</div>
</div>
</div>
<div id="tabset-7-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-7-2-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-21_e989624f946fe5a174af939efba75f43">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.heterozygosity</span>(data.gl, <span class="at">t.lower =</span> <span class="dv">0</span>, <span class="at">t.upper =</span> <span class="fl">0.3</span>,  <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.filter.heterozygosity 
Completed: :: 
 Completed: dartR.base 
 Completed: gl.filter.heterozygosity </code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="invariable-sites-1" class="level3">
<h3 class="anchored" data-anchor-id="invariable-sites-1">Invariable sites</h3>
<p>When filtering has resulted in removal of some individuals or populations, variation at several loci may be lost. Some loci may even be scored as missing across all individuals. You may wish to remove these monomorphic loci from your dataset.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-8-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-8-1" role="tab" aria-controls="tabset-8-1" aria-selected="true">Filter missing data</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-8-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-8-2" role="tab" aria-controls="tabset-8-2" aria-selected="false">Report monomorphs</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-8-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-8-3" role="tab" aria-controls="tabset-8-3" aria-selected="false">Filter monomorphs</a></li></ul>
<div class="tab-content">
<div id="tabset-8-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-8-1-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-22_3d7c12d191ebe4ee1f0cfba5aafadf07">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.allna</span>(data.gl, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.filter.allna 
Completed: :: 
 Completed: dartR.base 
 Completed: gl.filter.allna </code></pre>
</div>
</div>
</div>
<div id="tabset-8-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-8-2-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-23_88b44d1fe825a45811dc44050013e909">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.monomorphs</span>(data.gl, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.monomorphs 

  No. of loci: 2373 
    Polymorphic loci: 2373 
    Monomorphic loci: 0 
    Loci scored all NA: 1 
  No. of individuals: 460 
  No. of populations: 5 

Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.monomorphs </code></pre>
</div>
</div>
</div>
<div id="tabset-8-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-8-3-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-24_711258a091259c6e737972c0ce51cdca">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.monomorphs</span>(data.gl, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.filter.monomorphs 
Completed: :: 
 Completed: dartR.base 
 Completed: gl.filter.monomorphs </code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="linked-loci" class="level3">
<h3 class="anchored" data-anchor-id="linked-loci">Linked loci</h3>
<p>SNPs that are close together will be inherited together. Consequently, linked SNPs are not independent and will bias the data by inflating the signal.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Do not filter on LD, if you aim to analyse you data of effective population size based on linkage disequilibrium.</p>
</div>
</div>
<section id="short-distance-linkage" class="level4">
<h4 class="anchored" data-anchor-id="short-distance-linkage">Short-distance linkage</h4>
<p>Sequence tags can contain more than one callable SNP marker. Because of their close proximity, these multiple SNPs within a single sequence tag (referred to in dartR as <strong>‘secondaries’</strong> are likely to be strongly linked (inherited together). This is problematic for many analyses, so one might wish to filter out the multiple SNPs to leave only one per sequence tag.</p>
<p>This function removes SNPs on the same 70 bp sequence tag, it retains the SNP with the highest RepAvg (average reproducibility) andrAvgPIC (average polymorphic information content), respectively.</p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-9-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-9-1" role="tab" aria-controls="tabset-9-1" aria-selected="true">Report secondaries</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-9-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-9-2" role="tab" aria-controls="tabset-9-2" aria-selected="false">Filter secondaries</a></li></ul>
<div class="tab-content">
<div id="tabset-9-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-9-1-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-25_a8dab3be7f5e13a00e399962b5e16121">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.secondaries</span>(data.gl, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.secondaries </code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>  Total number of SNP loci scored: 2373 
   Number of sequence tags in total: 2216 
   Estimated number of invariant sequence tags: 15002 
   Number of sequence tags with secondaries: 145 
   Number of secondary SNP loci that would be removed on 
            filtering: 157 
   Number of SNP loci that would be retained on filtering: 2216 
   Number of invariant sites in sequenced tags: 134077 
   Mean length of sequence tags: 61.57491 
   Total Number of invariant sites (including invariant sequence 
            tags): 1057824 
Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.secondaries </code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>               Param        Value
1       n.total.tags 2.216000e+03
2 n.SNPs.secondaries 1.570000e+02
3   n.invariant.tags 1.500200e+04
4 n.tags.secondaries 1.450000e+02
5          n.inv.gen 1.340770e+05
6       mean.len.tag 6.157491e+01
7        n.invariant 1.057824e+06
8             Lambda 1.377714e-01</code></pre>
</div>
</div>
</div>
<div id="tabset-9-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-9-2-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-26_6eac2b4a8633f4d9b6e806712d3cb8f8">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.secondaries</span>(data.gl, <span class="at">method =</span> <span class="st">"best"</span>, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.filter.secondaries 
Completed: :: 
 Completed: dartR.base 
 Completed: gl.filter.secondaries </code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="long-distance-linkage" class="level4">
<h4 class="anchored" data-anchor-id="long-distance-linkage">Long-distance linkage</h4>
<p><strong>Not running this because it takes too long, but consider using this if you have many SNPs and a small genome.</strong></p>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-10-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-10-1" role="tab" aria-controls="tabset-10-1" aria-selected="true">Report ld.map</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-10-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-10-2" role="tab" aria-controls="tabset-10-2" aria-selected="false">Report ld</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-10-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-10-3" role="tab" aria-controls="tabset-10-3" aria-selected="false">Filter</a></li></ul>
<div class="tab-content">
<div id="tabset-10-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-10-1-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-27_894c7685093aa9a9ff504caae8b95055">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>ld.report <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.report.ld.map</span>(data.gl, <span class="at">ld.stat =</span> <span class="st">"R.squared"</span>, <span class="at">verbose =</span> <span class="dv">1</span>)</span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="co"># changes between ''.' and '_' depending on your version of snpStats</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>ld.report<span class="sc">$</span>ld_stat <span class="ot">&lt;-</span> ld.report<span class="sc">$</span>ld.stat </span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>ld.report<span class="sc">$</span>locus_a.stat_keep <span class="ot">&lt;-</span> ld.report<span class="sc">$</span>locus_a.stat.keep</span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>ld.report<span class="sc">$</span>locus_b.stat_keep <span class="ot">&lt;-</span> ld.report<span class="sc">$</span>locus_b.stat.keep</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-10-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-10-2-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-28_3792432297577b814e736af524130a21">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.ld</span>(data.gl, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div id="tabset-10-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-10-3-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-29_fe6f0817241eb0d05d1d83244370749c">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.ld</span>(data.gl, ld.report, <span class="at">threshold =</span> <span class="fl">0.25</span>, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</div>
</section>
</section>
<section id="hardy-weinberg-equilibrium" class="level3">
<h3 class="anchored" data-anchor-id="hardy-weinberg-equilibrium">Hardy-Weinberg Equilibrium</h3>
<p>The Hardy–Weinberg equilibrium (HWE) states that allele and genotype frequencies in a population will remain constant from generation to generation in the absence of other evolutionary influences. SNPs that are sampled from a population in demographic equilibrium are expected to follow this equation: <span class="math inline">\(p^2 + 2pq + q^2 = 1\)</span>, where <span class="math inline">\(p\)</span> is the frequency of the major allele (A) and <span class="math inline">\(q\)</span> is frequency of the minor allele (a).</p>
<p>In the simplest case of a single locus with two alleles denoted A and a with frequencies <span class="math inline">\(f(A) = p\)</span> and <span class="math inline">\(f(a) = q\)</span>, respectively, the expected genotype frequencies under random mating are <span class="math inline">\(f(AA) = p^2\)</span> for the AA homozygotes, <span class="math inline">\(f(aa) = q^2\)</span> for the aa homozygotes, and <span class="math inline">\(f(Aa) = 2pq\)</span> for the heterozygotes.</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>There are several factors that can cause deviations from Hardy-Weinberg proportions including:</p>
<ul>
<li>Mutations</li>
<li>Finite population size</li>
<li>Selection</li>
<li>Population structure</li>
<li>Age structure</li>
<li>Assortative mating</li>
<li>Sex linkage</li>
<li>Non-random sampling</li>
<li>Genotyping errors</li>
</ul>
<p>Therefore, testing for Hardy-Weinberg proportions should be a process that involves a careful evaluation of the results, a good place to start is <span class="citation" data-cites="waples_hwe">Waples (<a href="references.html#ref-waples_hwe" role="doc-biblioref">2015</a>)</span>. Note that tests for H-W proportions are only valid if there is no population substructure (assuming random mating) and have sufficient power only when there is sufficient sample size (n individuals &gt; 15).</p>
</div>
</div>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-11-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-11-1" role="tab" aria-controls="tabset-11-1" aria-selected="true">hwe diagnostics</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-11-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-11-2" role="tab" aria-controls="tabset-11-2" aria-selected="false">Report hwe</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-11-3-tab" data-bs-toggle="tab" data-bs-target="#tabset-11-3" role="tab" aria-controls="tabset-11-3" aria-selected="false">Filter hwe</a></li></ul>
<div class="tab-content">
<div id="tabset-11-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-11-1-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-30_1b8ef50d12aefb9c53ab115281f0bd7a">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.diagnostics.hwe</span>(data.gl, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.diagnostics.hwe 
Starting gl.colors 
Selected color type 2c 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_line()`: Each group consists of only one observation.
ℹ Do you need to adjust the group aesthetic?</code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-30-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.colors 
Selected color type 2 
Completed: gl.colors 
Starting gl.colors 
Selected color type 2c 
Completed: gl.colors 
Starting utils.jackknife 
  Processing genlight object with SNP data
The variation of Fis and Fst, respectively
 (measured as 
               standard error with the Jackknife method - see De Meeus 2018) 
               is: 2.10540973158501e-06, 1.65673151873847e-06 
 Fis vs Fst 
               ratio is: 1.27 </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 7 rows containing non-finite values (`stat_smooth()`).</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Removed 7 rows containing missing values (`geom_point()`).</code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-30-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>      Population nSig nExpected Deficiency Excess PropDeficiency ChiSquare
  Cambridge Gulf   55     110.8         40     15      0.7272727  2184.287
      Daly River   78     110.8         52     26      0.6666667  2233.706
      King Sound   42     110.8         14     28      0.3333333  1245.482
 Van Diemen Gulf  207     110.8        140     67      0.6763285  7903.792
        pvalue
  1.000000e+00
  1.000000e+00
  1.000000e+00
 7.619698e-200
Completed: :: 
 Completed: dartR.base 
 Completed: gl.diagnostics.hwe </code></pre>
</div>
</div>
</div>
<div id="tabset-11-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-11-2-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-31_93940cb01afbe76461c6897990783d89">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.hwe</span>(data.gl, <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.hwe 
Starting gl.colors 
Selected color type 2c 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_line()`: Each group consists of only one observation.
ℹ Do you need to adjust the group aesthetic?</code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-31-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>    Reporting significant departures from Hardy-Weinberg 
            Equilibrium
    NB: Departures significant at the alpha level of 0.05 are listed
    Adjustment of p-values for multiple comparisons vary
                    with sample size
         Population            Locus Hom_1 Het Hom_2   N         Prob Sig
    Van Diemen Gulf 100001551-62-C/T   153 153    72 378 3.550897e-03 sig
         Daly River 100001738-59-T/C    22   3     2  27 3.726019e-02 sig
         Daly River 100001760-48-G/A    23   3     2  28 3.460801e-02 sig
     Cambridge Gulf 100001906-15-C/T    22   5     3  30 3.228943e-02 sig
    Van Diemen Gulf 100002098-21-C/T   115 206    58 379 3.544120e-02 sig
---                                                                      
    Van Diemen Gulf   8171499-22-A/G     3 212   162 377 1.542178e-15 sig
     Cambridge Gulf   8171660-43-C/T     0  22     8  30 1.832795e-03 sig
         Daly River   8171660-43-C/T     0  24     4  28 8.156677e-05 sig
         King Sound   8171660-43-C/T     1  18     0  19 1.510845e-04 sig
    Van Diemen Gulf   8171660-43-C/T     8 267   103 378 4.829275e-26 sig
    Prob.adj Sig.adj npop
          NA      NA    1
          NA      NA    1
          NA      NA    1
          NA      NA    1
          NA      NA    1
---                      
          NA      NA    2
          NA      NA    4
          NA      NA    4
          NA      NA    4
          NA      NA    4

Completed: :: 
 
Completed: dartR.base 
 
Completed: gl.report.hwe </code></pre>
</div>
</div>
</div>
<div id="tabset-11-3" class="tab-pane" role="tabpanel" aria-labelledby="tabset-11-3-tab">
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-32_f9605d477b31e417bf81db5c3f444dcf">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a>data.gl <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.filter.hwe</span>(data.gl, <span class="at">n.pop.threshold =</span> <span class="dv">3</span>, </span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">mult.comp.adj =</span> <span class="cn">TRUE</span>, </span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">mult.comp.adj.method =</span> <span class="st">"BH"</span>, </span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>                                     <span class="at">verbose =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.filter.hwe 
Completed: :: 

 Completed: dartR.base 

 Completed: gl.filter.hwe </code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="remove-duplicate-individuals" class="level3">
<h3 class="anchored" data-anchor-id="remove-duplicate-individuals">Remove duplicate individuals</h3>
<p>Since DArT includes technical replicates and you could potentially recapture individuals, it important to remove replicate individuals. Duplicates individuals will inflate the signal you are investigating.</p>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-33_0f2b02c0876e8ea35731034ccd457653">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>res_dup <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.report.replicates</span>(data.gl,<span class="at">loc_threshold =</span> <span class="dv">500</span>, <span class="at">perc_geno =</span> <span class="fl">0.85</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.report.replicates 
  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Completed: :: 
 Completed: dartR.base 
 Completed: gl.report.replicates </code></pre>
</div>
</div>
</section>
</section>
<section id="filtering-conclusion" class="level2">
<h2 class="anchored" data-anchor-id="filtering-conclusion">Filtering Conclusion</h2>
<p><strong>Influence of Planned Analyses</strong></p>
<p>Possibly the most influential consideration on filtering is the purpose for which the filtered data set is to be used.</p>
<ul>
<li><p>If you are planning to generate <strong><em>high resolution linkage maps</em></strong>, then highly stringent filtering is warranted, and Diversity Arrays Technology would adjust their pipelines accordingly upon request. Similarly, if you are contemplating an analysis of <strong><em>relatedness or inbreeding</em></strong>, then stringent filtering might be warranted. In any case, it would be wise to start out with high stringency and then progressively relax the stringency and examine the impact on outcomes.</p></li>
<li><p>If your focus is on identifying <strong><em>sex-linked markers</em></strong>, which rely heavily on identifying markers that are heterozygous in XY individuals and homozygous in XX individuals, then a filtering threshold for read depth of 10x would be sensible. Sequencing volume in the context of genome size will provide an upper limit to how far you can push the read depth threshold, so in some cases, a high sequence volume service will need to be requested. Because the pipelines of Diversity Arrays Technology depend in part on Mendelian behaviour in the selection of SNP markers, and sex-linked loci do not behave in Mendelian fashion, you might also ask them to relax the stringency of some aspects of their filtering in generating the report.</p></li>
<li><p>If your focus is on <strong><em>Genome-wide Association Studies (GWAS)</em></strong> or identifying <strong><em>loci under selection</em></strong>, then noise in the data will not associate with phenotype, and so filtering can be less stringent in order to maximize the chances of identifying promising associations.</p></li>
<li><p><strong><em>Principal Coordinates Analysis</em></strong> PCA (and to a lesser extent, <strong><em>Principal Components Analysis</em></strong>, PCoA), rely on fully populated data matrices (no missing values). Classical PCA for example, cannot easily accommodate missing values. To overcome this, a balance must be struck between filtering on Call Rate and imputing the values of those missing values that remain. More stringent filtering on Call Rate, and less imputation; but more stringent filtering on Call Rate, more loss of potentially useful information or valuable samples. Refer to <span class="citation" data-cites="Georges2023">Georges et al. (<a href="references.html#ref-Georges2023" role="doc-biblioref">2023</a>)</span> for further discussion of this point.</p></li>
</ul>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>There is no hard and fast rule to guide decisions on filtering SNP datasets prior to a substantive analysis. The decisions are based on comparing the distribution of the parameters to be filtered (using one of the gl.report functions) and the purpose to which the filtered dataset is to be put. Sequencing volume can constrain options, and the likelihood of some level of contamination of samples or presence of inhibitors of the restriction enzymes can have a bearing on the decisions. It is important not to over filter because of the risk of introducing a level of systematic bias and because of the unnecessary loss of informative loci or individuals. An experimental approach is recommended, whereby different filtering regimes are tried (from stringent to less stringent) and checked for influence on analysis outcomes.</p>
</div>
</div>
</section>
</section>
<section id="population-structure" class="level1">
<h1>Population structure</h1>
<p>Inferring ancestry differences among individuals from different populations, or identifying population structure has been motivated by many applications:</p>
<ul>
<li><p>Population genetics</p></li>
<li><p>Genetic association studies</p></li>
<li><p>Prediction of whether an individual will have particular traits (for instance in people that have the duffy antigen that allows them to have resistance to malaria)</p></li>
</ul>
<p>Until recently we were only able to group individuals into populations using a few genetic markers. However, the decrease in cost and improvements in technology mean that is now feasible to use whole genomes to perform these kinds of analyses. This means that we can now have very fine scaled analyses of populations based on many thousands of markers.</p>
<p>Examining population structure can give us a great deal of insight into the history and origin of populations. Model-free methods for examining population structure and ancestry, such as <strong><em>principal components analysis</em></strong> are extremely popular in population genomic research. This is because it is typically simple to apply and relatively easy to interpret.</p>
<section id="principal-component-analysis" class="level2">
<h2 class="anchored" data-anchor-id="principal-component-analysis">Principal Component Analysis</h2>
<p>Principal components analysis (PCA) is the most commonly recognised and used method for identifying population structure. Essentially, PCA aims to identify the main axes of variation in a dataset with each axis being independent of the next (i.e.&nbsp;there should be no correlation between them). The first component summarizes the major axis variation and the second the next largest and so on, until cumulatively all the available variation is explained. In the context of genetic data, PCA summarizes the major axes of variation in allele frequencies and then produces the coordinates of individuals along these axes. For the rest of the tutorial we will conduct a PCA to demonstrate how it can help give insight to population structure within species.</p>
<p>This analysis allows us to calculate principal components (PCs) that explain the differences between individuals in genetic data. The naming of each of these components is based on the percentage of variation in the data they explain, with principal component 1 (PC1) being the most explanatory PC of the data. For instance if PC1 explains 5.7%. PC2 will have the next largest contribution to the genetic variance, for example PC2 may contribute 3.4%. The overall variance should add up to 100%.</p>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-34_86bf608eb2c78ba3d383022686ad5e3c">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>PCA <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.pcoa</span>(data.gl, <span class="at">nfactors =</span> <span class="dv">5</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.colors 
Selected color type 2 
Completed: gl.colors </code></pre>
</div>
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA, data.gl, <span class="at">xaxis =</span> <span class="dv">1</span>, <span class="at">yaxis =</span> <span class="dv">2</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-34-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.pcoa.plot</span>(PCA, data.gl, <span class="at">xaxis =</span> <span class="dv">2</span>, <span class="at">yaxis =</span> <span class="dv">3</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-34-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-34-4.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" id="id" class="class" width="48" height="48"></p>
<ul>
<li><p>How many genetic clusters are present across the sequenced samples? A cluster can be defined as a group of individuals that are genetically very similar to each other. Each cluster is genetically distinct from other clusters.</p></li>
<li><p>What is the relationship among these genetic clusters? i.e.&nbsp;which cluster is the most genetically distinct? Which genetic clusters are most similar to one another?</p></li>
<li><p>How many PCs is it worth plotting?</p></li>
<li><p>Can you identify the cluster membership of each individual? There are a variety of ways to do this. For example, you can look at the <code>PCA$scores</code> and sort the columns by PC.</p></li>
</ul>
</div>
</div>
</section>
<section id="fixation-indices" class="level2">
<h2 class="anchored" data-anchor-id="fixation-indices">Fixation indices</h2>
<p><span class="math inline">\(FST\)</span>, also known as the fixation index, is an extremely important statistic in population genetics, molecular ecology and evolutionary biology. It is also arguably one of the most famous population genetic statistics you will encounter. <span class="math inline">\(FST\)</span> essentially measures the level of genetic differentiation between two or more populations. It ranges from 0 (i.e.&nbsp;no genetic differentiation) to 1 (complete genetic differentiation)</p>
<p>Ultimately, it is quite a simple statistic to understand but it can sometimes take time to properly grasp. So we will go over the basics properly here. One of the most confusing things about <span class="math inline">\(FST\)</span> is that are several different ways to define it. For ease of understanding, we will use a simple formulation:</p>
<p><span class="math inline">\(FST=(HT−HS)/HT\)</span></p>
<p>For simplicity, imagine we are examining two populations only. With this formulate, <span class="math inline">\(HT\)</span> is the expected heterozygosity when the two populations are considered as one large meta-population. <span class="math inline">\(HS\)</span> is the average expected heterozygosity in the two populations.</p>
<p>You might be thinking, hang on a minute… what do we mean by expected heterozygosity? To appreciate this, we need to think back to the Hardy-Weinberg expectation we learned about in Section 4.3.9. Remember that at a simple bi-allelic locus, <span class="math inline">\(p\)</span> and <span class="math inline">\(q\)</span> are the frequencies of the two alleles. We can calculate the expected frequency of heterozygotes with <span class="math inline">\(2pq\)</span> - this is the expected heterozygosity.</p>
<p>You can use <span class="math inline">\(FST\)</span> to identify the relationship between individuals, and it can be used to identify populations. To do this, you first need to define the members of each of your populations. You did this at the end of the PCA section where you identified cluster membership.</p>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-35_f0f915fb80c1f38e6489e05b07b91ea3">
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a>fst.res <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.fst.pop</span>(data.gl, <span class="at">nboots =</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a>fst.res<span class="sc">$</span>Fsts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 Cambridge Gulf King Sound Daly River Van Diemen Gulf
Cambridge Gulf               NA         NA         NA              NA
King Sound           0.28898221         NA         NA              NA
Daly River           0.09406955  0.3033545         NA              NA
Van Diemen Gulf      0.11954153  0.2523445  0.0865569              NA
Papua New Guinea     0.15241388  0.3989535  0.1833683       0.1691328
                 Papua New Guinea
Cambridge Gulf                 NA
King Sound                     NA
Daly River                     NA
Van Diemen Gulf                NA
Papua New Guinea               NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>fst.res<span class="sc">$</span>Pvalues</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 Cambridge Gulf King Sound Daly River Van Diemen Gulf
Cambridge Gulf               NA         NA         NA              NA
King Sound                    0         NA         NA              NA
Daly River                    0          0         NA              NA
Van Diemen Gulf               0          0          0              NA
Papua New Guinea              0          0          0               0
                 Papua New Guinea
Cambridge Gulf                 NA
King Sound                     NA
Daly River                     NA
Van Diemen Gulf                NA
Papua New Guinea               NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="do">## correction for multiple testing: Bonferroni, Benjamini-Hochberg</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>Fst.p.adj <span class="ot">&lt;-</span> stats<span class="sc">::</span><span class="fu">p.adjust</span>(fst.res<span class="sc">$</span>Pvalues, <span class="st">"BH"</span>)</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>Fst.p.adj <span class="ot">&lt;-</span> <span class="fu">matrix</span>(Fst.p.adj, <span class="at">ncol =</span> adegenet<span class="sc">::</span><span class="fu">nPop</span>(data.gl))</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(Fst.p.adj) <span class="ot">&lt;-</span> adegenet<span class="sc">::</span><span class="fu">popNames</span>(data.gl)</span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(Fst.p.adj) <span class="ot">&lt;-</span> adegenet<span class="sc">::</span><span class="fu">popNames</span>(data.gl)</span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a>Fst.p.adj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 Cambridge Gulf Daly River King Sound Papua New Guinea
Cambridge Gulf               NA         NA         NA               NA
Daly River                    0         NA         NA               NA
King Sound                    0          0         NA               NA
Papua New Guinea              0          0          0               NA
Van Diemen Gulf               0          0          0                0
                 Van Diemen Gulf
Cambridge Gulf                NA
Daly River                    NA
King Sound                    NA
Papua New Guinea              NA
Van Diemen Gulf               NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>fixed.diff <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.fixed.diff</span>(data.gl, <span class="at">tloc =</span> <span class="fl">0.1</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>fixed.diff<span class="sc">$</span>fd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 Cambridge Gulf Daly River King Sound Papua New Guinea
Daly River                    0                                       
King Sound                    9         15                            
Papua New Guinea              6          3         23                 
Van Diemen Gulf               0          0          5                4</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" class="class" width="48" height="48"></p>
<ul>
<li><p>Do the <span class="math inline">\(FST\)</span> metrics support the PCA results?</p></li>
<li><p>Do you see a different significance in p-values when correcting for multiple comparisons?</p></li>
</ul>
</div>
</div>
</section>
</section>
<section id="genetic-diversity" class="level1">
<h1>Genetic diversity</h1>
<p>Genetic variability of populations of a species is important because it can be thought of an indirect measure of the evolvability of those populations. The greater the genetic diversity of a population the bigger the amount of genetic raw material upon which selection can act. A population deprived of its genetic variability is less likely to respond to environmental changes and more vulnerable to the negative effects of inbreeding. From a theoretical point of you measuring (correctly) genetic diversity of a population and comparing it among different naturally occurring or human driven treatments might be important to test the theoretical predictions on impact of drift on effect of population size of relatively isolated (small) populations and on mechanisms of genetic variance (diversity) maintenance. From the conservation point of view the relative genetic diversity of population differentially affected by human intervention might give insight into how resilient populations of a species are to rapid changes of the habitat in which they live.</p>
<ul>
<li><p><strong>Ho = Observed heterozygosity</strong></p>
<p><span class="math inline">\(Ho= 1-\sum_k \sum_i Pkii/np\)</span></p>
<p>where <span class="math inline">\(Pkii\)</span> represents the proportion of homozygote <span class="math inline">\(i\)</span> in sample <span class="math inline">\(k\)</span> and <span class="math inline">\(np\)</span> the number of samples.</p></li>
<li><p><strong>Hs = Within population gene diversity (sometimes misleadingly called expected heterozygosity, He)</strong></p>
<p><span class="math inline">\(Hs=\tilde{n}/(\tilde{n}-1)[1-\sum_i\bar{p_i^2}-Ho/2\tilde{n}]\)</span></p>
<p>where <span class="math inline">\(\tilde{n}=np/\sum_k 1/n_k\)</span> and <span class="math inline">\(\bar{p_i^2}=\sum_k p_{ki}^2/np\)</span></p></li>
<li><p><strong>Ht = Overall gene diversity</strong></p>
<p><span class="math inline">\(Ht= 1-\sum_i\bar{p_i}^2+Hs/(\tilde{n} np)-Ho/(2\tilde{n}np)\)</span></p>
<p>where <span class="math inline">\(\bar{p_i}=\sum_kp_{ki}/np\)</span>.</p></li>
<li><p><strong>Dst = Amount of gene diversity among samples</strong></p>
<p><span class="math inline">\(Dst=Ht-Hs\)</span></p></li>
<li><p><strong>Htp = corrected Ht</strong></p>
<p><span class="math inline">\(Ht'=Hs+Dst'\)</span></p></li>
<li><p><strong>Dstp = Corrected Dst</strong></p>
<p><span class="math inline">\(Dst'=np/(np-1)Dst\)</span></p></li>
<li><p><strong>Fst = Fixation index</strong></p>
<p><span class="math inline">\(Fst=Dst/Ht\)</span> (This is not the same as Nei’s <span class="math inline">\(Gst\)</span>)</p></li>
<li><p><strong>Fstp = Corrected Fst</strong></p>
<p><span class="math inline">\(Fst'=Dst'/Ht'\)</span></p></li>
<li><p><strong>Fis = Inbreeding coefficient</strong></p>
<p><span class="math inline">\(Fis=1-Ho/Hs\)</span></p></li>
<li><p><strong>Dest = a measure of population differentiation as defined by <span class="citation" data-cites="jost2008gst">Jost (<a href="references.html#ref-jost2008gst" role="doc-biblioref">2008</a>)</span></strong></p>
<p><span class="math inline">\(Dest=np/(np-1) (Ht'-Hs)/(1-Hs)\)</span></p>
<p>where <span class="math inline">\(p_{ki}\)</span> are unweighted by sample size.</p></li>
<li><p><strong>Gst = an estimator of <span class="math inline">\(Fst\)</span> based on allele frequencies only</strong></p></li>
</ul>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-36_06e8afadb430f58deae0fd2b714edca1">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>basic.stats <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">utils.basic.stats</span>(data.gl)</span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>basic.stats<span class="sc">$</span>overall</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Ho      Hs      Ht     Dst     Htp    Dstp     Fst    Fstp     Fis    Dest 
 0.2176  0.2155  0.2610  0.0455  0.2724  0.0569  0.1743  0.2088 -0.0098  0.0725 
Gst_max   Gst_H 
 0.7506  0.2782 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colMeans</span>(basic.stats<span class="sc">$</span>Ho, <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="co">#observed heterozygosity</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Cambridge Gulf       Daly River       King Sound Papua New Guinea 
       0.2333104        0.2293272        0.1446049        0.2412934 
 Van Diemen Gulf 
       0.2391721 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>hets <span class="ot">&lt;-</span> <span class="fu">gl.report.heterozygosity</span>(data.gl, <span class="at">method =</span> <span class="st">"pop"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl.report.heterozygosity 
  Processing genlight object with SNP data
  Calculating Observed Heterozygosities, averaged across
                    loci, for each population
  Calculating Expected Heterozygosities

Starting gl.colors 
Selected color type dis 
Completed: gl.colors </code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>                              pop      n.Ind n.Loc n.Loc.adj polyLoc monoLoc
Cambridge Gulf     Cambridge Gulf  29.785262  2212         1    1752     460
Daly River             Daly River  27.773056  2212         1    1757     455
King Sound             King Sound  18.815956  2206         1    1048    1158
Papua New Guinea Papua New Guinea   3.956581  2211         1    1205    1006
Van Diemen Gulf   Van Diemen Gulf 376.841772  2212         1    2193      19
                 all_NALoc       Ho     HoSD     HoSE       He     HeSD
Cambridge Gulf           0 0.233314 0.202012 0.004295 0.231541 0.190550
Daly River               0 0.229330 0.202057 0.004296 0.227633 0.189611
King Sound               6 0.144606 0.212348       NA 0.134674 0.185966
Papua New Guinea         1 0.241294 0.281102       NA 0.204037 0.203145
Van Diemen Gulf          0 0.239172 0.188658 0.004011 0.240018 0.186304
                     HeSE      uHe    uHeSD    uHeSE       FIS    FISSD FISSE
Cambridge Gulf   0.004052 0.235494 0.193803 0.004121  0.010577 0.185612    NA
Daly River       0.004032 0.231807 0.193087 0.004105  0.009830 0.194138    NA
King Sound             NA 0.138351 0.191042       NA -0.029191 0.220399    NA
Papua New Guinea       NA 0.233551 0.232530       NA -0.026071 0.373230    NA
Van Diemen Gulf  0.003961 0.240337 0.186551 0.003966  0.010817 0.106491    NA
Completed: gl.report.heterozygosity </code></pre>
</div>
</div>
<p><strong>Some alternate measures for genetic diversity are more comparable between datasets.This is provided by the <code>gl.report.diversity</code> function.</strong></p>
<p>The output of this function: provide information for different indices: allelic richness (q = 0), Shannon information (q = 1), and heterozygosity (q = 2); within (alpha) or between (beta) populations. For all indexes, the entropies (H) and Hill numbers (D) are calculated.</p>
<p>In a nutshell, the alpha indexes between the different q-values should be similar if there is no deviation from expected allele frequencies and occurrences (e.g.&nbsp;all loci in HWE &amp; equilibrium). If there is a deviation of an index, this links to a process causing it, such as dispersal, selection or strong drift. For a detailed explanation of all the indexes see <span class="citation" data-cites="sherwin">Sherwin et al. (<a href="references.html#ref-sherwin" role="doc-biblioref">2017</a>)</span>.</p>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-37_3cda663efbb9f748d1e1b965d24b9771">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>divers <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.report.diversity</span>(data.gl, <span class="at">table =</span> <span class="cn">FALSE</span>, <span class="at">pbar =</span> <span class="cn">FALSE</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>divers<span class="sc">$</span>one_H_alpha</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Cambridge Gulf       Daly River       King Sound Papua New Guinea 
       0.3529367        0.3482662        0.2051896        0.3026173 
 Van Diemen Gulf 
       0.3687394 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>divers<span class="sc">$</span>one_H_beta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 Cambridge Gulf Daly River King Sound Papua New Guinea
Cambridge Gulf               NA 0.14604195 0.20724507       0.20825765
Daly River           0.03065942         NA 0.19289681       0.19368266
King Sound           0.10234696 0.10471342         NA       0.27058669
Papua New Guinea     0.05352050 0.05579714 0.12753651               NA
Van Diemen Gulf      0.02042284 0.02275808 0.09423184       0.04555016
                 Van Diemen Gulf
Cambridge Gulf        0.08479532
Daly River            0.07111530
King Sound            0.16381330
Papua New Guinea      0.16576583
Van Diemen Gulf               NA</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" class="class" width="48" height="48"></p>
<ul>
<li><p>Can you compare observed versus expected heterozygosity per population? Why would they be different?</p></li>
<li><p>Which population has the lowest genetic diversity? What would this mean for the population?</p></li>
<li><p>Is inbreeding (FIS) prominent in any population? What could cause this?</p></li>
</ul>
</div>
</div>
</section>
<section id="population-connectivity" class="level1">
<h1>Population connectivity</h1>
<p><code>divMigrate</code>&nbsp;uses the method described in <span class="citation" data-cites="sundqvist_divmigrate">Sundqvist et al. (<a href="references.html#ref-sundqvist_divmigrate" role="doc-biblioref">2016</a>)</span> to plot the relative migration levels between population samples.</p>
<p>Nm (i.e., the effective number of migrants). This statistic incorporates complementary information from both Gst and D, suggesting it may be a more generally suitable measure of migration. When using Nm for calculating the percent of correct directions in the different simulation scenarios.</p>
<p>Alternatively, the number of private alleles provides an indication of genetic isolation. We use private alleles (i.e.&nbsp;alleles only present in one population) as a sign of no migration.</p>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-38_998fb00c7b17b0f54167899188339447">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"gl2gpop_FDD.R"</span>)</span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a><span class="fu">gl2gpop</span>(data.gl, <span class="at">filename =</span> <span class="st">"data/genepop.gen"</span>)</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a><span class="co"># dartR.base::gl2genepop(data.gl, outfile = "genepop.gen", outpath = getwd(), verbose = 5) # Not recognised by diveRsity</span></span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a>migration <span class="ot">&lt;-</span> diveRsity<span class="sc">::</span><span class="fu">divMigrate</span>(<span class="at">infile =</span> <span class="st">"data/genepop.gen"</span>, <span class="at">boots =</span> <span class="dv">100</span>, <span class="at">stat =</span> <span class="st">"Nm"</span>, <span class="at">plot_network =</span> <span class="cn">TRUE</span>, <span class="at">para =</span> <span class="cn">TRUE</span>, <span class="at">filter_threshold =</span> <span class="fl">0.2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-38-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(migration<span class="sc">$</span>nmRelMig) <span class="ot">&lt;-</span> adegenet<span class="sc">::</span><span class="fu">popNames</span>(data.gl)</span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(migration<span class="sc">$</span>nmRelMig) <span class="ot">&lt;-</span> adegenet<span class="sc">::</span><span class="fu">popNames</span>(data.gl)</span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a>migration<span class="sc">$</span>nmRelMig</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                 Cambridge Gulf Daly River King Sound Papua New Guinea
Cambridge Gulf               NA  0.7857795 0.10183280        0.2322329
Daly River            0.8169289         NA 0.09510506        0.1900063
King Sound            0.2703799  0.2472216         NA        0.1117764
Papua New Guinea      0.3508070  0.3018515 0.07791770               NA
Van Diemen Gulf       0.5989464  0.9129015 0.10757361        0.2043979
                 Van Diemen Gulf
Cambridge Gulf         0.6963752
Daly River             1.0000000
King Sound             0.3130946
Papua New Guinea       0.3640824
Van Diemen Gulf               NA</code></pre>
</div>
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Private alleles</span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a>dartR.base<span class="sc">::</span><span class="fu">gl.report.pa</span>(data.gl, <span class="at">plot.display =</span> <span class="cn">TRUE</span>, <span class="at">verbose =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Processing genlight object with SNP data</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   p1 p2             pop1             pop2 N1  N2 fixed priv1 priv2 Chao1 Chao2
1   1  2   Cambridge Gulf       Daly River 30  28     0   211   216     0     0
2   1  3   Cambridge Gulf       King Sound 30  19     0   833   129     0     0
3   1  4   Cambridge Gulf Papua New Guinea 30   4     1   627    80     0    12
4   1  5   Cambridge Gulf  Van Diemen Gulf 30 379     0    17   458     0     0
5   2  3       Daly River       King Sound 28  19     0   843   134     0     0
6   2  4       Daly River Papua New Guinea 28   4     1   634    82     0     8
7   2  5       Daly River  Van Diemen Gulf 28 379     0    10   446     0     0
8   3  4       King Sound Papua New Guinea 19   4    15   377   535     1     8
9   3  5       King Sound  Van Diemen Gulf 19 379     1     5  1151     0     0
10  4  5 Papua New Guinea  Van Diemen Gulf  4 379     0     7   995     0     0
   totalpriv   AFD asym asym.sig
1        427 0.115   NA       NA
2        962 0.197   NA       NA
3        707 0.166   NA       NA
4        475 0.130   NA       NA
5        977 0.200   NA       NA
6        716 0.179   NA       NA
7        456 0.109   NA       NA
8        912 0.225   NA       NA
9       1156 0.198   NA       NA
10      1002 0.175   NA       NA</code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p><img src="images/task.png" class="class" width="48" height="48"></p>
<ul>
<li><p>Which populations are highly connected?</p></li>
<li><p>How does this concur with knowledge on the dispersal ability and historical seascape of the species?</p></li>
<li><p>Does your connectivity changes if you use a different statistics (e.g.&nbsp;‘D’ or ‘gst’)?</p></li>
</ul>
</div>
</div>
</section>
<section id="natural-selection-in-the-context-of-adaptive-divergence" class="level1">
<h1>Natural selection in the context of adaptive divergence</h1>
<p>Understanding natural selection is fundamental to properly understanding evolution. The basics of natural selection are relatively straightforward but surprisingly easily misunderstood. Our aim with this practical session is to use the R environment to actually demonstrate different methods to detect signals of selection.</p>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>Both analyses to identify natural selection are much stronger if you have reference genome for your species, so you can pinpoint the genes and their functions that are being selected.</p>
</div>
</div>
<section id="fst-based-outliers" class="level2">
<h2 class="anchored" data-anchor-id="fst-based-outliers"><span class="math inline">\(FST\)</span>-based outliers</h2>
<p><strong>Good to detect strong selective pressure</strong></p>
<p>This method looks for <span class="math inline">\(FST\)</span> outliers from a list of <span class="math inline">\(FST\)</span>’s for different loci. It assumes that each locus has been genotyped in all populations with approximately equal coverage. Here, we identify loci ‘putatively’ under selection per population using the <strong><em>outflank</em></strong> method of <span class="citation" data-cites="whitlock_outliers">Whitlock and Lotterhos (<a href="references.html#ref-whitlock_outliers" role="doc-biblioref">2015</a>)</span>.</p>
<p>OutFLANK estimates the distribution of <span class="math inline">\(FST\)</span> based on a trimmed sample of <span class="math inline">\(FST\)</span>’s. It assumes that the majority of loci in the center of the distribution are neutral and infers the shape of the distribution of neutral <span class="math inline">\(FST\)</span> using a trimmed set of loci. Loci with the highest and lowest <span class="math inline">\(FST\)</span>’s are trimmed from the data set before this inference, and the distribution of <span class="math inline">\(FST df/(mean FST)\)</span> is assumed to follow a chi-square distribution. Based on this inferred distribution, each locus is given a q-value based on its quantile in the inferred null distribution.</p>
<p>More information here:</p>
<p><a href="https://popgen.nescent.org/2016-01-26-SNP-selection.html">Detection of the signal of selection from genome scan</a></p>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-39_5d4852385d96bbb8465aa5144ea6d92f">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>outliers <span class="ot">&lt;-</span> dartR.popgen<span class="sc">::</span><span class="fu">gl.outflank</span>(data.gl,</span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">plot =</span> <span class="cn">TRUE</span>,</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">LeftTrimFraction =</span> <span class="fl">0.05</span>,</span>
<span id="cb112-4"><a href="#cb112-4" aria-hidden="true" tabindex="-1"></a>                        <span class="at">RightTrimFraction =</span> <span class="fl">0.05</span>,</span>
<span id="cb112-5"><a href="#cb112-5" aria-hidden="true" tabindex="-1"></a>                        <span class="at">Hmin =</span> <span class="fl">0.1</span>,</span>
<span id="cb112-6"><a href="#cb112-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">qthreshold =</span> <span class="fl">0.05</span>,)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting gl2gi 
  Processing genlight object with SNP data
Matrix converted.. Prepare genind object...
Completed: gl2gi 
Calculating FSTs, may take a few minutes...</code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>outliers<span class="sc">$</span>outflank<span class="sc">$</span>numberLowFstOutliers</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a>outliers<span class="sc">$</span>outflank<span class="sc">$</span>numberHighFstOutliers</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
</div>
</section>
<section id="genotype-environment-associations" class="level2">
<h2 class="anchored" data-anchor-id="genotype-environment-associations">Genotype-Environment associations</h2>
<p><strong>Better at detecting weak selective pressures</strong></p>
<p>Here, we illustrate the use of <strong><em>Redundancy Analysis (RDA)</em></strong> as a genotype-environment association (GEA) method to detect loci under selection <span class="citation" data-cites="forester_gea">(<a href="references.html#ref-forester_gea" role="doc-biblioref">Forester et al. 2018</a>)</span>. RDA is a multivariate ordination technique that can be used to analyze many loci and environmental predictors simultaneously. RDA determines how groups of loci co-vary in response to the multivariate environment, and can detect processes that result in weak, multilocus molecular signatures <span class="citation" data-cites="forester_gea">(<a href="references.html#ref-forester_gea" role="doc-biblioref">Forester et al. 2018</a>)</span>.</p>
<p>More information here: <a href="https://popgen.nescent.org/2018-03-27_RDA_GEA.html">Detecting multilocus adaptation using Redundancy Analysis (RDA)</a></p>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-40_aa9d77e4d50307bd00fd5dd4bcd2692e">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a>data.gl.imputed <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl.impute</span>(data.gl, <span class="at">parallel =</span> <span class="cn">FALSE</span>) <span class="co"># no missing data allowed</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Starting :: 
 Starting dartR.base 
 Starting gl.impute 
  Processing genlight object with SNP data
  Imputation based on drawing from the nearest neighbour
  Residual missing values were filled randomly drawing from the global allele profiles by locus
Completed: :: 
 Completed: dartR.base 
 Completed: gl.impute </code></pre>
</div>
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>gen <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(data.gl.imputed)</span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>predictor <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(data.gl.imputed<span class="sc">@</span>other<span class="sc">$</span>ind.metrics)</span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a>psych<span class="sc">::</span><span class="fu">pairs.panels</span>(predictor[,<span class="dv">2</span><span class="sc">:</span><span class="dv">4</span>], <span class="at">scale =</span> T, <span class="at">cex.cor =</span> <span class="dv">3</span>) <span class="co">#to check correlation between pairs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For simplicity, we take the population as a proxy for environment (and sex a s a covariate), but ideally you have fine-scale environmental data for each individual. </span></span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>SNP_rda <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">rda</span>(gen <span class="sc">~</span> pop <span class="sc">+</span> <span class="fu">Condition</span>(sex), </span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a>                      <span class="at">data =</span> predictor, <span class="at">scale =</span> <span class="cn">TRUE</span>) </span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(SNP_rda)<span class="sc">$</span>cont<span class="sc">$</span>importance[,<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                             RDA1        RDA2        RDA3         RDA4
Eigenvalue            94.49748776 59.59813062 25.31846964 12.886197785
Proportion Explained   0.04317671  0.02723089  0.01156822  0.005887813
Cumulative Proportion  0.04317671  0.07040760  0.08197582  0.087863636</code></pre>
</div>
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(vegan<span class="sc">::</span><span class="fu">eigenvals</span>(SNP_rda, <span class="at">model =</span> <span class="st">"constrained"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Importance of components:
                         RDA1    RDA2    RDA3     RDA4
Eigenvalue            94.4975 59.5981 25.3185 12.88620
Proportion Explained   0.4914  0.3099  0.1317  0.06701
Cumulative Proportion  0.4914  0.8013  0.9330  1.00000</code></pre>
</div>
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We have as many constrained ("RDA") axes as we have predictors </span></span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Residual variance is then modelled by PCA (the unconstrained "PC" axes)</span></span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Variance explained by the environmental predictors is under "Proportion"</span></span>
<span id="cb125-4"><a href="#cb125-4" aria-hidden="true" tabindex="-1"></a><span class="co"># and "Constrained" = equivalent to R2 in multiple regression,</span></span>
<span id="cb125-5"><a href="#cb125-5" aria-hidden="true" tabindex="-1"></a><span class="co"># so it is biased &amp; should be adjusted based on the # of predictors:</span></span>
<span id="cb125-6"><a href="#cb125-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb125-7"><a href="#cb125-7" aria-hidden="true" tabindex="-1"></a>vegan<span class="sc">::</span><span class="fu">RsquareAdj</span>(SNP_rda)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$r.squared
[1] 0.08693503

$adj.r.squared
[1] 0.07931155</code></pre>
</div>
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Our constrained ordination explains about 6% of the variation</span></span>
<span id="cb127-2"><a href="#cb127-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Not surprising given our large # of SNPs and other processes at work </span></span>
<span id="cb127-3"><a href="#cb127-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb127-4"><a href="#cb127-4" aria-hidden="true" tabindex="-1"></a>SNP.aov <span class="ot">&lt;-</span> <span class="fu">anova</span>(SNP_rda, <span class="at">test =</span> <span class="st">"Chi"</span>, <span class="at">permutations =</span> <span class="dv">100</span>, <span class="at">parallel =</span> <span class="dv">4</span>) <span class="co">#run Anova on the result to see the significance level</span></span>
<span id="cb127-5"><a href="#cb127-5" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(SNP.aov)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Permutation test for rda under reduced model
Permutation: free
Number of permutations: 100

Model: rda(formula = gen ~ pop + Condition(sex), data = predictor, scale = TRUE)
          Df Variance      F   Pr(&gt;F)   
Model      4    192.3 10.909 0.009901 **
Residual 453   1996.3                   
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Finally let's check Variance Inflation Factors for our predictors (they should be ok)</span></span>
<span id="cb129-2"><a href="#cb129-2" aria-hidden="true" tabindex="-1"></a>vegan<span class="sc">::</span><span class="fu">vif.cca</span>(SNP_rda)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               sexM                sexU       popDaly River       popKing Sound 
           1.044034            2.272482            1.816568            1.612040 
popPapua New Guinea  popVan Diemen Gulf 
           2.400668            2.415785 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="co"># All are below 10 and most are below 5 = multicollinearity shouldn't be a problem</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a><span class="fu">screeplot</span>(SNP_rda, <span class="at">npcs =</span> <span class="dv">20</span>, <span class="at">ptype =</span> <span class="st">"lines"</span>, <span class="at">bstick =</span> F)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-40-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a>pop <span class="ot">&lt;-</span> predictor<span class="sc">$</span>pop</span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a>col <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"#ff7f00"</span>,<span class="st">"#1f78b4"</span>,<span class="st">"#ffff33"</span>,<span class="st">"#a6cee3"</span>,<span class="st">"#33a02c"</span>,<span class="st">"#e31a1c"</span>)</span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(SNP_rda, <span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), <span class="at">type =</span> <span class="st">"n"</span>, <span class="at">scaling =</span> <span class="dv">3</span>)</span>
<span id="cb132-4"><a href="#cb132-4" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(SNP_rda, <span class="at">display =</span> <span class="st">"species"</span>, <span class="at">pch =</span> <span class="dv">20</span>, <span class="at">cex =</span> <span class="fl">0.7</span>, <span class="at">col =</span> <span class="st">"gray32"</span>, <span class="at">scaling =</span> <span class="dv">3</span>)           <span class="co"># the SNPs</span></span>
<span id="cb132-5"><a href="#cb132-5" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(SNP_rda, <span class="at">display =</span> <span class="st">"sites"</span>, <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">cex =</span> <span class="fl">1.3</span>, <span class="at">col =</span> <span class="st">"gray32"</span>, <span class="at">scaling =</span> <span class="dv">3</span>, <span class="at">bg =</span> col[pop]) <span class="co"># the individuals</span></span>
<span id="cb132-6"><a href="#cb132-6" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(SNP_rda, <span class="at">scaling =</span> <span class="dv">3</span>, <span class="at">display =</span> <span class="st">"bp"</span>, <span class="at">col =</span> <span class="st">"#0868ac"</span>, <span class="at">cex =</span> <span class="dv">1</span>)                           <span class="co"># the predictors</span></span>
<span id="cb132-7"><a href="#cb132-7" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"bottomright"</span>, <span class="at">legend =</span> <span class="fu">levels</span>(pop), <span class="at">bty =</span> <span class="st">"n"</span>, <span class="at">col =</span> <span class="st">"gray32"</span>, <span class="at">pch =</span> <span class="dv">21</span>, <span class="at">cex =</span> <span class="dv">1</span>, <span class="at">pt.bg =</span> col)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-40-3.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify candidates</span></span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a><span class="co"># -------------------</span></span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a>load.rda <span class="ot">&lt;-</span> vegan<span class="sc">::</span><span class="fu">scores</span>(SNP_rda, <span class="at">choices =</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>), <span class="at">display =</span> <span class="st">"species"</span>)  <span class="co"># Species scores for the first three constrained axes</span></span>
<span id="cb133-4"><a href="#cb133-4" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(load.rda[,<span class="dv">1</span>], <span class="at">main =</span> <span class="st">"Loadings on RDA1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-40-4.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(load.rda[,<span class="dv">2</span>], <span class="at">main =</span> <span class="st">"Loadings on RDA2"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-40-5.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This simple function identifies the SNPs that load in the tails of these distributions. </span></span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Outliers are defines as 3 standard deviation cut-off (two-tailed p-value = 0.0027).</span></span>
<span id="cb135-3"><a href="#cb135-3" aria-hidden="true" tabindex="-1"></a>outliers <span class="ot">&lt;-</span> <span class="cf">function</span>(x,z){</span>
<span id="cb135-4"><a href="#cb135-4" aria-hidden="true" tabindex="-1"></a>  lims <span class="ot">&lt;-</span> <span class="fu">mean</span>(x) <span class="sc">+</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">1</span>) <span class="sc">*</span> z <span class="sc">*</span> <span class="fu">sd</span>(x)     <span class="co"># find loadings +/-z sd from mean loading</span></span>
<span id="cb135-5"><a href="#cb135-5" aria-hidden="true" tabindex="-1"></a>  x[x <span class="sc">&lt;</span> lims[<span class="dv">1</span>] <span class="sc">|</span> x <span class="sc">&gt;</span> lims[<span class="dv">2</span>]]               <span class="co"># locus names in these tails</span></span>
<span id="cb135-6"><a href="#cb135-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb135-7"><a href="#cb135-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb135-8"><a href="#cb135-8" aria-hidden="true" tabindex="-1"></a>cand1 <span class="ot">&lt;-</span> <span class="fu">outliers</span>(load.rda[,<span class="dv">1</span>],<span class="dv">3</span>)</span>
<span id="cb135-9"><a href="#cb135-9" aria-hidden="true" tabindex="-1"></a>cand2 <span class="ot">&lt;-</span> <span class="fu">outliers</span>(load.rda[,<span class="dv">2</span>],<span class="dv">3</span>)</span>
<span id="cb135-10"><a href="#cb135-10" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">length</span>(cand1) <span class="sc">+</span> <span class="fu">length</span>(cand2))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 30</code></pre>
</div>
</div>
</section>
</section>
<section id="relatedness" class="level1">
<h1>Relatedness</h1>
<p>Two alleles are identical by descent (IBD) if they are identical copies of the same ancestral allele in a base population. Relatedness is the proportion of alleles shared between individuals that are IBD. Kinship is the probability that two alleles, one taken at random from each individual, are IBD <span class="citation" data-cites="wang2022_kinship">(<a href="references.html#ref-wang2022_kinship" role="doc-biblioref">Wang 2022</a>)</span>. Kinship considers half the genetic information because it looks at the probability that one individual inherits a particular allele from a common ancestor shared with another individual. Therefore, kinship is equal to half of relatedness.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>Below is a table modified from <span class="citation" data-cites="speed2015_relatedness">Speed and Balding (<a href="references.html#ref-speed2015_relatedness" role="doc-biblioref">2015</a>)</span> showing kinship coefficients (<span class="math inline">\(\theta\)</span>), and their confidence intervals (CI), and relatedness values for different relationships.</p>
<table class="table">
<colgroup>
<col style="width: 54%">
<col style="width: 12%">
<col style="width: 21%">
<col style="width: 12%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Relationship</th>
<th style="text-align: center;">Kinship</th>
<th style="text-align: center;">CI(95%)</th>
<th style="text-align: center;">r</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Identical twins/clones/same individual</td>
<td style="text-align: center;">0.5</td>
<td style="text-align: center;">-</td>
<td style="text-align: center;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">Sibling/Parent-Offspring</td>
<td style="text-align: center;">0.25</td>
<td style="text-align: center;">(0.204, 0.296)</td>
<td style="text-align: center;">0.5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Half-sibling</td>
<td style="text-align: center;">0.125</td>
<td style="text-align: center;">(0.092, 0.158)</td>
<td style="text-align: center;">0.25</td>
</tr>
<tr class="even">
<td style="text-align: left;">First cousin</td>
<td style="text-align: center;">0.062</td>
<td style="text-align: center;">(0.038, 0.089)</td>
<td style="text-align: center;">0.125</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Half-cousin</td>
<td style="text-align: center;">0.031</td>
<td style="text-align: center;">(0.012, 0.055)</td>
<td style="text-align: center;">0.062</td>
</tr>
<tr class="even">
<td style="text-align: left;">Second cousin</td>
<td style="text-align: center;">0.016</td>
<td style="text-align: center;">(0.004, 0.031)</td>
<td style="text-align: center;">0.031</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Half-second cousin</td>
<td style="text-align: center;">0.008</td>
<td style="text-align: center;">(0.001, 0.020)</td>
<td style="text-align: center;">0.016</td>
</tr>
<tr class="even">
<td style="text-align: left;">Third cousin</td>
<td style="text-align: center;">0.004</td>
<td style="text-align: center;">(0.000, 0.012)</td>
<td style="text-align: center;">0.008</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Unrelated</td>
<td style="text-align: center;">0</td>
<td style="text-align: center;">-</td>
<td style="text-align: center;">0</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Note that the variance of relatedness depends on the number of SNPs and genome size, therefore even with perfect data your relatedness will not follow these expected values exactly. <img src="images/r_distribution.png" class="img-fluid"></p>
<p>Further, in species with small population sizes (not uncommon in Chondrichthyes), inbreeding may skew the expected values even more. <img src="images/r_FSFS.png" class="img-fluid"></p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Relatedness/kinship are not absolute states but are estimated relative to a reference population for which there is generally little information, so that we can estimate the kinship of a pair of individuals only relative to some other quantity.</p>
<p>All relatedness estimates assume the true allele frequencies of your natural populations are known. But <strong>small samples size, missing data, DNA contamination, populations structure, inbreeding, and the presence of related individuals</strong> in your data will bias your frequency distributions.</p>
<p>Previous studies showed that around 5,000 SNPs are sufficient to accurately estimate relatedness and kinship <span class="citation" data-cites="goudet_kiship">(<a href="references.html#ref-goudet_kiship" role="doc-biblioref">Goudet, Kay, and Weir 2018</a>)</span>.</p>
</div>
</div>
<section id="convert-data" class="level2">
<h2 class="anchored" data-anchor-id="convert-data">Convert data</h2>
<p>There are many good software available to estimate relatedness, such as COLONY2, sequoia, gcta64, EMIBD9, PC-Relate, etc. For this example, we use the <code>related</code> package <span class="citation" data-cites="pew_related">(<a href="references.html#ref-pew_related" role="doc-biblioref">Pew et al. 2015</a>)</span>.</p>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-41_973d640b0698d3a811475c21d97e9faf">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>relate <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl2related</span>(data.gl, <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a>pop.code <span class="ot">&lt;-</span> <span class="fu">substr</span>(data.gl<span class="sc">$</span>pop, <span class="at">start =</span> <span class="dv">1</span>, <span class="at">stop =</span> <span class="dv">2</span>) <span class="co">#Related only takes 2 letter code for population</span></span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a>relate<span class="sc">$</span>V1 <span class="ot">&lt;-</span> <span class="fu">paste0</span>(pop.code, <span class="st">"_"</span>, relate<span class="sc">$</span>V1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-coancestry-on-full-data" class="level2">
<h2 class="anchored" data-anchor-id="run-coancestry-on-full-data">Run coancestry on full data</h2>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-42_68d4a4aec168c6250e3deba3440a08bb">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>pw.rel <span class="ot">&lt;-</span> related<span class="sc">::</span><span class="fu">coancestry</span>(relate, <span class="at">lynchli =</span> <span class="dv">1</span>, <span class="at">lynchrd =</span> <span class="dv">1</span>,<span class="at">trioml =</span> 0L, <span class="at">dyadml =</span> 0L,</span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a>                               <span class="at">quellergt =</span> <span class="dv">1</span>, <span class="at">wang =</span> <span class="dv">1</span>, <span class="at">ritland =</span> <span class="dv">1</span>)</span>
<span id="cb138-3"><a href="#cb138-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb138-4"><a href="#cb138-4" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(pw.rel, <span class="at">file =</span> <span class="st">"data/relatedness.Rdata"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-43_1875ace105dbb6359830b834f1d8fd6c">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">load</span>(<span class="st">"data/relatedness.Rdata"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pw.rel"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> pw.rel<span class="sc">$</span>relatedness</span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-3"><a href="#cb141-3" aria-hidden="true" tabindex="-1"></a>plot.data <span class="ot">&lt;-</span> tidyr<span class="sc">::</span><span class="fu">gather</span>(res, condition, measurement, <span class="dv">5</span><span class="sc">:</span><span class="dv">11</span>, <span class="at">factor_key =</span> <span class="cn">TRUE</span>)</span>
<span id="cb141-4"><a href="#cb141-4" aria-hidden="true" tabindex="-1"></a>summ <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">group_by</span>(res, group) <span class="sc">%&gt;%</span></span>
<span id="cb141-5"><a href="#cb141-5" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">summarise</span>(<span class="fu">mean</span>(ritland))</span>
<span id="cb141-6"><a href="#cb141-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-7"><a href="#cb141-7" aria-hidden="true" tabindex="-1"></a>kableExtra<span class="sc">::</span><span class="fu">kbl</span>(summ) <span class="sc">%&gt;%</span></span>
<span id="cb141-8"><a href="#cb141-8" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">kable_styling</span>(<span class="at">full_width =</span> <span class="cn">FALSE</span>, <span class="at">font_size =</span> <span class="dv">11</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">group</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">mean(ritland)</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">CaCa</td>
<td style="text-align: right;">0.2748690</td>
</tr>
<tr class="even">
<td style="text-align: left;">CaDa</td>
<td style="text-align: right;">0.0889671</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CaKi</td>
<td style="text-align: right;">0.0637159</td>
</tr>
<tr class="even">
<td style="text-align: left;">CaPa</td>
<td style="text-align: right;">0.1612474</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CaVa</td>
<td style="text-align: right;">-0.0362530</td>
</tr>
<tr class="even">
<td style="text-align: left;">DaCa</td>
<td style="text-align: right;">0.0885335</td>
</tr>
<tr class="odd">
<td style="text-align: left;">DaDa</td>
<td style="text-align: right;">0.1585196</td>
</tr>
<tr class="even">
<td style="text-align: left;">DaKi</td>
<td style="text-align: right;">0.0128862</td>
</tr>
<tr class="odd">
<td style="text-align: left;">DaPa</td>
<td style="text-align: right;">0.0302889</td>
</tr>
<tr class="even">
<td style="text-align: left;">DaVa</td>
<td style="text-align: right;">-0.0202242</td>
</tr>
<tr class="odd">
<td style="text-align: left;">KiCa</td>
<td style="text-align: right;">0.0695242</td>
</tr>
<tr class="even">
<td style="text-align: left;">KiDa</td>
<td style="text-align: right;">0.0177524</td>
</tr>
<tr class="odd">
<td style="text-align: left;">KiKi</td>
<td style="text-align: right;">0.6588158</td>
</tr>
<tr class="even">
<td style="text-align: left;">KiPa</td>
<td style="text-align: right;">0.0913111</td>
</tr>
<tr class="odd">
<td style="text-align: left;">KiVa</td>
<td style="text-align: right;">-0.0420271</td>
</tr>
<tr class="even">
<td style="text-align: left;">PaCa</td>
<td style="text-align: right;">0.1595705</td>
</tr>
<tr class="odd">
<td style="text-align: left;">PaDa</td>
<td style="text-align: right;">0.0475053</td>
</tr>
<tr class="even">
<td style="text-align: left;">PaKi</td>
<td style="text-align: right;">0.0897825</td>
</tr>
<tr class="odd">
<td style="text-align: left;">PaPa</td>
<td style="text-align: right;">1.0283667</td>
</tr>
<tr class="even">
<td style="text-align: left;">PaVa</td>
<td style="text-align: right;">-0.0383714</td>
</tr>
<tr class="odd">
<td style="text-align: left;">VaCa</td>
<td style="text-align: right;">-0.0364779</td>
</tr>
<tr class="even">
<td style="text-align: left;">VaDa</td>
<td style="text-align: right;">-0.0247332</td>
</tr>
<tr class="odd">
<td style="text-align: left;">VaKi</td>
<td style="text-align: right;">-0.0431017</td>
</tr>
<tr class="even">
<td style="text-align: left;">VaPa</td>
<td style="text-align: right;">-0.0309888</td>
</tr>
<tr class="odd">
<td style="text-align: left;">VaVa</td>
<td style="text-align: right;">0.0044785</td>
</tr>
</tbody>
</table>


</div>
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a>plot <span class="ot">&lt;-</span> ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(plot.data[plot.data<span class="sc">$</span>condition <span class="sc">==</span> <span class="st">"wang"</span>,], </span>
<span id="cb142-2"><a href="#cb142-2" aria-hidden="true" tabindex="-1"></a>                        ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> group, <span class="at">y =</span> measurement, </span>
<span id="cb142-3"><a href="#cb142-3" aria-hidden="true" tabindex="-1"></a>                                                <span class="at">fill =</span> group)) <span class="sc">+</span> </span>
<span id="cb142-4"><a href="#cb142-4" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_boxplot</span>() <span class="sc">+</span></span>
<span id="cb142-5"><a href="#cb142-5" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">facet_wrap</span>(condition<span class="sc">~</span>.) <span class="sc">+</span></span>
<span id="cb142-6"><a href="#cb142-6" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Site"</span>, <span class="at">y =</span> <span class="st">"Relatedness"</span>) <span class="sc">+</span></span>
<span id="cb142-7"><a href="#cb142-7" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust =</span> <span class="dv">1</span>))</span>
<span id="cb142-8"><a href="#cb142-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(plot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-43-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb143"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb143-1"><a href="#cb143-1" aria-hidden="true" tabindex="-1"></a>hist.plot <span class="ot">&lt;-</span> ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(plot.data[plot.data<span class="sc">$</span>measurement <span class="sc">&gt;</span> <span class="fl">0.1</span> <span class="sc">&amp;</span> </span>
<span id="cb143-2"><a href="#cb143-2" aria-hidden="true" tabindex="-1"></a>                        plot.data<span class="sc">$</span>group <span class="sc">==</span> <span class="st">"VaVa"</span> <span class="sc">&amp;</span> </span>
<span id="cb143-3"><a href="#cb143-3" aria-hidden="true" tabindex="-1"></a>                        plot.data<span class="sc">$</span>condition <span class="sc">==</span> <span class="st">"wang"</span>,],</span>
<span id="cb143-4"><a href="#cb143-4" aria-hidden="true" tabindex="-1"></a>                        ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> measurement)) <span class="sc">+</span></span>
<span id="cb143-5"><a href="#cb143-5" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_histogram</span>(<span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">fill =</span> <span class="st">"lightgray"</span>, <span class="at">bins =</span>  <span class="dv">100</span>) <span class="sc">+</span></span>
<span id="cb143-6"><a href="#cb143-6" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Wang - relatedness"</span>,</span>
<span id="cb143-7"><a href="#cb143-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">y =</span> <span class="st">"Frequency"</span>)</span>
<span id="cb143-8"><a href="#cb143-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hist.plot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-43-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise
</div>
</div>
<div class="callout-body-container callout-body">
<p>Why do you have negative relatedness?</p>
<p>Why do some populations have very high relatedness?</p>
</div>
</div>
</section>
<section id="one-population" class="level2">
<h2 class="anchored" data-anchor-id="one-population">One population</h2>
<p>In reality, relatedness should only be estimated with a single population.</p>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-44_f461925ec2ed48f21486eb278a66e68f">
<div class="sourceCode cell-code" id="cb144"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb144-1"><a href="#cb144-1" aria-hidden="true" tabindex="-1"></a>popname <span class="ot">&lt;-</span> <span class="st">"Van Diemen Gulf"</span></span>
<span id="cb144-2"><a href="#cb144-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb144-3"><a href="#cb144-3" aria-hidden="true" tabindex="-1"></a>relate<span class="fl">.1</span>pop <span class="ot">&lt;-</span> dartR.base<span class="sc">::</span><span class="fu">gl2related</span>(data.gl[data.gl<span class="sc">$</span>pop <span class="sc">==</span> popname,], <span class="at">verbose =</span> <span class="dv">0</span>)</span>
<span id="cb144-4"><a href="#cb144-4" aria-hidden="true" tabindex="-1"></a>pop.code <span class="ot">&lt;-</span> <span class="st">"Va"</span> <span class="co">#Related only takes 2 letter code for population</span></span>
<span id="cb144-5"><a href="#cb144-5" aria-hidden="true" tabindex="-1"></a>relate<span class="fl">.1</span>pop<span class="sc">$</span>V1 <span class="ot">&lt;-</span> <span class="fu">paste0</span>(pop.code, <span class="st">"_"</span>, relate<span class="fl">.1</span>pop<span class="sc">$</span>V1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-45_237ba9fe83fccca17addf19587ea6d12">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a>pw.rel<span class="fl">.1</span>pop <span class="ot">&lt;-</span> related<span class="sc">::</span><span class="fu">coancestry</span>(relate<span class="fl">.1</span>pop, <span class="at">lynchli =</span> <span class="dv">1</span>, <span class="at">lynchrd =</span> <span class="dv">1</span>,<span class="at">trioml =</span> 0L, <span class="at">dyadml =</span> 0L,</span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a>                               <span class="at">quellergt =</span> <span class="dv">1</span>, <span class="at">wang =</span> <span class="dv">1</span>, <span class="at">ritland =</span> <span class="dv">1</span>)</span>
<span id="cb145-3"><a href="#cb145-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb145-4"><a href="#cb145-4" aria-hidden="true" tabindex="-1"></a><span class="fu">save</span>(pw.rel<span class="fl">.1</span>pop, <span class="at">file =</span> <span class="st">"data/relatedness_1pop.Rdata"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-46_800686291c871ea84b8936439f6b5e3c">
<div class="sourceCode cell-code" id="cb146"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb146-1"><a href="#cb146-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">load</span>(<span class="st">"data/relatedness_1pop.Rdata"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pw.rel.1pop"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a>res<span class="fl">.1</span>pop <span class="ot">&lt;-</span> pw.rel<span class="fl">.1</span>pop<span class="sc">$</span>relatedness</span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a>plot.data<span class="fl">.1</span>pop <span class="ot">&lt;-</span> tidyr<span class="sc">::</span><span class="fu">gather</span>(res<span class="fl">.1</span>pop, condition, measurement, <span class="dv">5</span><span class="sc">:</span><span class="dv">11</span>, <span class="at">factor_key =</span> <span class="cn">TRUE</span>)</span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb148-4"><a href="#cb148-4" aria-hidden="true" tabindex="-1"></a>hist.plot <span class="ot">&lt;-</span> ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(plot.data<span class="fl">.1</span>pop[plot.data<span class="fl">.1</span>pop<span class="sc">$</span>measurement <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> </span>
<span id="cb148-5"><a href="#cb148-5" aria-hidden="true" tabindex="-1"></a>                        plot.data<span class="fl">.1</span>pop<span class="sc">$</span>condition <span class="sc">==</span> <span class="st">"wang"</span>,],</span>
<span id="cb148-6"><a href="#cb148-6" aria-hidden="true" tabindex="-1"></a>                        ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> measurement)) <span class="sc">+</span></span>
<span id="cb148-7"><a href="#cb148-7" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_histogram</span>(<span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">fill =</span> <span class="st">"lightgray"</span>, <span class="at">bins =</span>  <span class="dv">100</span>) <span class="sc">+</span></span>
<span id="cb148-8"><a href="#cb148-8" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Wang - relatedness"</span>,</span>
<span id="cb148-9"><a href="#cb148-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">y =</span> <span class="st">"Frequency"</span>)</span>
<span id="cb148-10"><a href="#cb148-10" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hist.plot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-46-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb149"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb149-1"><a href="#cb149-1" aria-hidden="true" tabindex="-1"></a>high.rel <span class="ot">&lt;-</span> plot.data<span class="fl">.1</span>pop[plot.data<span class="fl">.1</span>pop<span class="sc">$</span>measurement <span class="sc">&gt;</span> <span class="fl">0.2</span> <span class="sc">&amp;</span> </span>
<span id="cb149-2"><a href="#cb149-2" aria-hidden="true" tabindex="-1"></a>                        plot.data<span class="fl">.1</span>pop<span class="sc">$</span>condition <span class="sc">==</span> <span class="st">"wang"</span>,]</span>
<span id="cb149-3"><a href="#cb149-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb149-4"><a href="#cb149-4" aria-hidden="true" tabindex="-1"></a>hist.plot <span class="ot">&lt;-</span> ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(high.rel, ggplot2<span class="sc">::</span><span class="fu">aes</span>(<span class="at">x =</span> measurement)) <span class="sc">+</span></span>
<span id="cb149-5"><a href="#cb149-5" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_histogram</span>(<span class="at">colour =</span> <span class="st">"black"</span>, <span class="at">fill =</span> <span class="st">"lightgray"</span>, <span class="at">bins =</span>  <span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb149-6"><a href="#cb149-6" aria-hidden="true" tabindex="-1"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Wang - relatedness"</span>,</span>
<span id="cb149-7"><a href="#cb149-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">y =</span> <span class="st">"Frequency"</span>)</span>
<span id="cb149-8"><a href="#cb149-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(hist.plot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-46-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a>kableExtra<span class="sc">::</span><span class="fu">kbl</span>(high.rel[<span class="dv">1</span><span class="sc">:</span><span class="dv">20</span>, <span class="fu">c</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">6</span>)], </span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">longtable =</span> <span class="cn">TRUE</span>, <span class="at">booktabs =</span> <span class="cn">TRUE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a>  kableExtra<span class="sc">::</span><span class="fu">kable_styling</span>(<span class="at">full_width =</span> <span class="cn">FALSE</span>,<span class="at">font_size =</span> <span class="dv">9</span>, </span>
<span id="cb150-4"><a href="#cb150-4" aria-hidden="true" tabindex="-1"></a>                            <span class="at">latex_options =</span> <span class="fu">c</span>(<span class="st">"repeat_header"</span>, <span class="st">"hold_position"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th"></th>
<th style="text-align: left;" data-quarto-table-cell-role="th">ind1.id</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">ind2.id</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">group</th>
<th style="text-align: left;" data-quarto-table-cell-role="th">condition</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">measurement</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">71634</td>
<td style="text-align: left;">Va_GGA353</td>
<td style="text-align: left;">Va_GGA345</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.4183</td>
</tr>
<tr class="even">
<td style="text-align: left;">71836</td>
<td style="text-align: left;">Va_GGA353</td>
<td style="text-align: left;">Va_GGA364</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.2373</td>
</tr>
<tr class="odd">
<td style="text-align: left;">71838</td>
<td style="text-align: left;">Va_GGA353</td>
<td style="text-align: left;">Va_GGA366</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.2816</td>
</tr>
<tr class="even">
<td style="text-align: left;">71991</td>
<td style="text-align: left;">Va_GGA353</td>
<td style="text-align: left;">Va_GGA358</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.2977</td>
</tr>
<tr class="odd">
<td style="text-align: left;">71993</td>
<td style="text-align: left;">Va_GGA353</td>
<td style="text-align: left;">Va_GGA360</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.2242</td>
</tr>
<tr class="even">
<td style="text-align: left;">72681</td>
<td style="text-align: left;">Va_GGL002</td>
<td style="text-align: left;">Va_GGL033</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.3874</td>
</tr>
<tr class="odd">
<td style="text-align: left;">72964</td>
<td style="text-align: left;">Va_GGA345</td>
<td style="text-align: left;">Va_GGA364</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.3010</td>
</tr>
<tr class="even">
<td style="text-align: left;">72965</td>
<td style="text-align: left;">Va_GGA345</td>
<td style="text-align: left;">Va_GGA365</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.2231</td>
</tr>
<tr class="odd">
<td style="text-align: left;">72966</td>
<td style="text-align: left;">Va_GGA345</td>
<td style="text-align: left;">Va_GGA366</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.3212</td>
</tr>
<tr class="even">
<td style="text-align: left;">73108</td>
<td style="text-align: left;">Va_GGA345</td>
<td style="text-align: left;">Va_GGA355</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.2462</td>
</tr>
<tr class="odd">
<td style="text-align: left;">73119</td>
<td style="text-align: left;">Va_GGA345</td>
<td style="text-align: left;">Va_GGA358</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.3050</td>
</tr>
<tr class="even">
<td style="text-align: left;">73575</td>
<td style="text-align: left;">Va_GGA347</td>
<td style="text-align: left;">Va_GGA088</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.5037</td>
</tr>
<tr class="odd">
<td style="text-align: left;">74902</td>
<td style="text-align: left;">Va_GGA350</td>
<td style="text-align: left;">Va_GGA369</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.2847</td>
</tr>
<tr class="even">
<td style="text-align: left;">75350</td>
<td style="text-align: left;">Va_GGA351</td>
<td style="text-align: left;">Va_GGA360</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.2675</td>
</tr>
<tr class="odd">
<td style="text-align: left;">76090</td>
<td style="text-align: left;">Va_GGL003</td>
<td style="text-align: left;">Va_GGL013</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.2712</td>
</tr>
<tr class="even">
<td style="text-align: left;">76093</td>
<td style="text-align: left;">Va_GGL003</td>
<td style="text-align: left;">Va_GGL016</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.2240</td>
</tr>
<tr class="odd">
<td style="text-align: left;">76716</td>
<td style="text-align: left;">Va_GGS004</td>
<td style="text-align: left;">Va_GGA206</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.2392</td>
</tr>
<tr class="even">
<td style="text-align: left;">76725</td>
<td style="text-align: left;">Va_GGS004</td>
<td style="text-align: left;">Va_GGA342</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.2146</td>
</tr>
<tr class="odd">
<td style="text-align: left;">76829</td>
<td style="text-align: left;">Va_GGS004</td>
<td style="text-align: left;">Va_GGS012</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.5121</td>
</tr>
<tr class="even">
<td style="text-align: left;">78226</td>
<td style="text-align: left;">Va_GGL006</td>
<td style="text-align: left;">Va_GGA111</td>
<td style="text-align: left;">VaVa</td>
<td style="text-align: left;">wang</td>
<td style="text-align: right;">0.2975</td>
</tr>
</tbody>
</table>


</div>
</div>
</section>
</section>
<section id="sex-linked-markers" class="level1">
<h1>Sex-linked markers</h1>
<p>See <span class="citation" data-cites="robledo_slm">Robledo-Ruiz et al. (<a href="references.html#ref-robledo_slm" role="doc-biblioref">2023</a>)</span> and <span class="citation" data-cites="devloo_slm">Devloo-Delva, Gosselin, et al. (<a href="references.html#ref-devloo_slm" role="doc-biblioref">2023</a>)</span>.</p>
<section id="load-data" class="level2">
<h2 class="anchored" data-anchor-id="load-data">Load data</h2>
<section id="bull-shark-1" class="level3">
<h3 class="anchored" data-anchor-id="bull-shark-1">Bull Shark</h3>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-47_8a4d5a39c2ec82cbca8b49c428095107">
<div class="sourceCode cell-code" id="cb151"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb151-1"><a href="#cb151-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">load</span>(<span class="st">"data/Bull_shark_DArTseq_genlight_for_Sex-linked_markers.Rdata"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.gl"</code></pre>
</div>
</div>
</section>
<section id="blue-shark-1" class="level3">
<h3 class="anchored" data-anchor-id="blue-shark-1">Blue Shark</h3>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-48_ea1c1ed4f2c57738c98ad79d29314128">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">load</span>(<span class="st">"data/Blue_shark_DArTseq_genlight_for_sex-linked_markers.Rdata"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.gl"</code></pre>
</div>
</div>
</section>
<section id="run-gl.filter.sexlinked" class="level3">
<h3 class="anchored" data-anchor-id="run-gl.filter.sexlinked">Run <code>gl.filter.sexlinked</code></h3>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/XY.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Markers on X and Y chromosomes</figcaption>
</figure>
</div>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-49_b0ce9b03fa3f9bb1fdd30a2aa2bfad3f">
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">gl.filter.sexlinked</span>(<span class="at">gl =</span> data.gl, <span class="at">system =</span> <span class="st">"xy"</span>, <span class="at">plots =</span> <span class="cn">TRUE</span>, <span class="at">ncores =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Detected 104 females and 111 males.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Starting phase 1. May take a while...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building call rate plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-49-1.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done. Starting phase 2.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Building heterozygosity plots.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-49-2.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-49-3.png" class="img-fluid" width="672"></p>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Done building heterozygosity plots.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>**FINISHED** Total of analyzed loci: 172384.
Found 26 sex-linked loci:
   2 Y-linked loci
   2 sex-biased loci
   22 X-linked loci
   0 XY gametologs.
And 172358 autosomal loci.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="main_code_files/figure-html/unnamed-chunk-49-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="run-gl.infer.sex" class="level3">
<h3 class="anchored" data-anchor-id="run-gl.infer.sex">Run <code>gl.infer.sex</code></h3>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-50_8d23a4138fd5e9457e49c14a50dd4348">
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a>sexID <span class="ot">&lt;-</span> dartR.sexlinked<span class="sc">::</span><span class="fu">gl.infer.sex</span>(<span class="at">gl_sex_filtered =</span> res, <span class="at">system =</span> <span class="st">"xy"</span>, <span class="at">seed =</span> <span class="dv">124</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Not enough gametologs (need at least 5). Assigning NA...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>***FINISHED***</code></pre>
</div>
<div class="sourceCode cell-code" id="cb166"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb166-1"><a href="#cb166-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">kable</span>(<span class="fu">head</span>(sexID))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 12%">
<col style="width: 7%">
<col style="width: 8%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 12%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 10%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: left;">id</th>
<th style="text-align: left;">y.linked.sex</th>
<th style="text-align: right;">#called</th>
<th style="text-align: right;">#missing</th>
<th style="text-align: left;">x.linked.sex</th>
<th style="text-align: right;">#Het.x</th>
<th style="text-align: right;">#Hom.x</th>
<th style="text-align: left;">gametolog.sex</th>
<th style="text-align: left;">#Het.g</th>
<th style="text-align: left;">#Hom.g</th>
<th style="text-align: left;">agreed.sex</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">60088</td>
<td style="text-align: left;">60088</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60160</td>
<td style="text-align: left;">60160</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">21</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">60168</td>
<td style="text-align: left;">60168</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60176</td>
<td style="text-align: left;">60176</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">21</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="odd">
<td style="text-align: left;">60096</td>
<td style="text-align: left;">60096</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">22</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
<tr class="even">
<td style="text-align: left;">60104</td>
<td style="text-align: left;">60104</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">0</td>
<td style="text-align: left;">M</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">19</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">NA</td>
<td style="text-align: left;">M</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
</section>
</section>
<section id="session-information" class="level1">
<h1>Session information</h1>
<div class="cell" data-hash="main_code_cache/html/unnamed-chunk-51_2aeadc17f445c9eb15bf055a741cce89">
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">width =</span> <span class="dv">100</span>)</span>
<span id="cb167-2"><a href="#cb167-2" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">session_info</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>─ Session info ───────────────────────────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.2.1 (2022-06-23 ucrt)
 os       Windows 10 x64 (build 22000)
 system   x86_64, mingw32
 ui       RTerm
 language (EN)
 collate  English_Australia.utf8
 ctype    English_Australia.utf8
 tz       Australia/Hobart
 date     2024-03-24
 pandoc   3.1.1 @ C:/Program Files/RStudio/resources/app/bin/quarto/bin/tools/ (via rmarkdown)

─ Packages ───────────────────────────────────────────────────────────────────────────────────────
 package         * version    date (UTC) lib source
 abind             1.4-5      2016-07-21 [1] CRAN (R 4.2.0)
 ade4            * 1.7-22     2023-02-06 [1] CRAN (R 4.2.3)
 adegenet        * 2.1.10     2023-01-26 [1] CRAN (R 4.2.3)
 ape               5.7-1      2023-03-13 [1] CRAN (R 4.2.3)
 backports         1.4.1      2021-12-13 [1] CRAN (R 4.2.0)
 base64enc         0.1-3      2015-07-28 [1] CRAN (R 4.2.0)
 bayesm            3.1-6      2023-09-23 [1] CRAN (R 4.2.3)
 BiocGenerics      0.44.0     2022-11-01 [1] Bioconductor
 cachem            1.0.8      2023-05-01 [1] CRAN (R 4.2.3)
 checkmate         2.3.1      2023-12-04 [1] CRAN (R 4.2.3)
 cli               3.6.2      2023-12-11 [1] CRAN (R 4.2.3)
 cluster           2.1.6      2023-12-01 [1] CRAN (R 4.2.3)
 codetools         0.2-19     2023-02-01 [1] CRAN (R 4.2.2)
 colorspace        2.1-0      2023-01-23 [1] CRAN (R 4.2.3)
 compositions      2.0-7      2024-01-09 [1] CRAN (R 4.2.3)
 corpcor           1.6.10     2021-09-16 [1] CRAN (R 4.2.0)
 crayon            1.5.2      2022-09-29 [1] CRAN (R 4.2.3)
 dartR.base      * 0.88       2024-03-22 [1] Github (green-striped-gecko/dartR.base@cfc4d92)
 dartR.data      * 1.0.2      2022-11-16 [1] CRAN (R 4.2.3)
 dartR.popgen    * 0.88       2024-03-22 [1] Github (green-striped-gecko/dartR.popgen@d21d9f3)
 dartR.sexlinked * 0.88       2024-03-22 [1] Github (green-striped-gecko/dartR.sexlinked@af6e6a2)
 dartRverse      * 0.51       2023-12-01 [1] CRAN (R 4.2.3)
 data.table        1.14.10    2023-12-08 [1] CRAN (R 4.2.3)
 DEoptimR          1.1-3      2023-10-07 [1] CRAN (R 4.2.3)
 devtools          2.4.5      2022-10-11 [1] CRAN (R 4.2.3)
 digest            0.6.34     2024-01-11 [1] CRAN (R 4.2.3)
 diveRsity       * 1.9.90     2024-01-19 [1] Github (kkeenan02/diveRsity@17d00ad)
 doParallel        1.0.17     2022-02-07 [1] CRAN (R 4.2.1)
 dplyr           * 1.1.4      2023-11-17 [1] CRAN (R 4.2.3)
 ellipsis          0.3.2      2021-04-29 [1] CRAN (R 4.2.1)
 evaluate          0.23       2023-11-01 [1] CRAN (R 4.2.3)
 fansi             1.0.6      2023-12-08 [1] CRAN (R 4.2.3)
 farver            2.1.1      2022-07-06 [1] CRAN (R 4.2.1)
 fastmap           1.1.1      2023-02-24 [1] CRAN (R 4.2.3)
 fdrtool           1.2.17     2021-11-13 [1] CRAN (R 4.2.0)
 forcats         * 1.0.0      2023-01-29 [1] CRAN (R 4.2.3)
 foreach           1.5.2      2022-02-02 [1] CRAN (R 4.2.1)
 foreign           0.8-86     2023-11-28 [1] CRAN (R 4.2.3)
 formatR         * 1.14       2023-01-17 [1] CRAN (R 4.2.3)
 Formula           1.2-5      2023-02-24 [1] CRAN (R 4.2.2)
 fs                1.6.3      2023-07-20 [1] CRAN (R 4.2.3)
 gdsfmt            1.38.1     2024-01-25 [1] Github (zhengxwen/gdsfmt@89f2673)
 generics          0.1.3      2022-07-05 [1] CRAN (R 4.2.1)
 ggdendro          0.2.0      2024-02-23 [1] CRAN (R 4.2.3)
 ggplot2         * 3.4.4      2023-10-12 [1] CRAN (R 4.2.3)
 ggtern          * 3.3.5      2021-07-23 [1] Github (mijangos81/ggtern@23fa82a)
 glasso            1.11       2019-10-01 [1] CRAN (R 4.2.0)
 glue              1.7.0      2024-01-09 [1] CRAN (R 4.2.3)
 gridExtra         2.3        2017-09-09 [1] CRAN (R 4.2.1)
 gtable            0.3.4      2023-08-21 [1] CRAN (R 4.2.3)
 gtools            3.9.5      2023-11-20 [1] CRAN (R 4.2.3)
 highr             0.10       2022-12-22 [1] CRAN (R 4.2.3)
 Hmisc             5.1-1      2023-09-12 [1] CRAN (R 4.2.3)
 hms               1.1.3      2023-03-21 [1] CRAN (R 4.2.3)
 htmlTable         2.4.2      2023-10-29 [1] CRAN (R 4.2.3)
 htmltools         0.5.7      2023-11-03 [1] CRAN (R 4.2.3)
 htmlwidgets       1.6.4      2023-12-06 [1] CRAN (R 4.2.3)
 httpuv            1.6.13     2023-12-06 [1] CRAN (R 4.2.3)
 igraph            1.6.0      2023-12-11 [1] CRAN (R 4.2.3)
 iterators         1.0.14     2022-02-05 [1] CRAN (R 4.2.1)
 jpeg              0.1-10     2022-11-29 [1] CRAN (R 4.2.2)
 jsonlite          1.8.8      2023-12-04 [1] CRAN (R 4.2.3)
 kableExtra        1.4.0      2024-01-24 [1] CRAN (R 4.2.1)
 knitr           * 1.45       2023-10-30 [1] CRAN (R 4.2.3)
 labeling          0.4.3      2023-08-29 [1] CRAN (R 4.2.3)
 later             1.3.2      2023-12-06 [1] CRAN (R 4.2.3)
 latex2exp         0.9.6      2022-11-28 [1] CRAN (R 4.2.3)
 lattice           0.22-5     2023-10-24 [1] CRAN (R 4.2.3)
 lavaan            0.6-17     2023-12-20 [1] CRAN (R 4.2.3)
 lifecycle         1.0.4      2023-11-07 [1] CRAN (R 4.2.3)
 lubridate       * 1.9.3      2023-09-27 [1] CRAN (R 4.2.3)
 magrittr          2.0.3      2022-03-30 [1] CRAN (R 4.2.1)
 MASS              7.3-60.0.1 2024-01-13 [1] CRAN (R 4.2.3)
 Matrix          * 1.6-5      2024-01-11 [1] CRAN (R 4.2.3)
 memoise           2.0.1      2021-11-26 [1] CRAN (R 4.2.3)
 mgcv              1.9-1      2023-12-21 [1] CRAN (R 4.2.3)
 mime              0.12       2021-09-28 [1] CRAN (R 4.2.0)
 miniUI            0.1.1.1    2018-05-18 [1] CRAN (R 4.2.3)
 mnormt            2.1.1      2022-09-26 [1] CRAN (R 4.2.1)
 munsell           0.5.0      2018-06-12 [1] CRAN (R 4.2.1)
 networkD3         0.4        2017-03-18 [1] CRAN (R 4.2.3)
 nlme              3.1-164    2023-11-27 [1] CRAN (R 4.2.3)
 nnet              7.3-19     2023-05-03 [1] CRAN (R 4.2.3)
 patchwork         1.2.0      2024-01-08 [1] CRAN (R 4.2.3)
 pbapply           1.7-2      2023-06-27 [1] CRAN (R 4.2.3)
 pbivnorm          0.6.0      2015-01-23 [1] CRAN (R 4.2.0)
 pegas             1.3        2023-12-13 [1] CRAN (R 4.2.3)
 permute           0.9-7      2022-01-27 [1] CRAN (R 4.2.1)
 pillar            1.9.0      2023-03-22 [1] CRAN (R 4.2.3)
 pkgbuild          1.4.3      2023-12-10 [1] CRAN (R 4.2.3)
 pkgconfig         2.0.3      2019-09-22 [1] CRAN (R 4.2.1)
 pkgload           1.3.4      2024-01-16 [1] CRAN (R 4.2.3)
 plyr              1.8.9      2023-10-02 [1] CRAN (R 4.2.3)
 png               0.1-8      2022-11-29 [1] CRAN (R 4.2.2)
 profvis           0.3.8      2023-05-02 [1] CRAN (R 4.2.3)
 promises          1.2.1      2023-08-10 [1] CRAN (R 4.2.3)
 proto             1.0.0      2016-10-29 [1] CRAN (R 4.2.3)
 psych             2.4.1      2024-01-18 [1] CRAN (R 4.2.1)
 purrr           * 1.0.2      2023-08-10 [1] CRAN (R 4.2.3)
 qgraph            1.9.8      2023-11-03 [1] CRAN (R 4.2.3)
 quadprog          1.5-8      2019-11-20 [1] CRAN (R 4.2.0)
 qvalue            2.30.0     2022-11-01 [1] Bioconductor
 R6                2.5.1      2021-08-19 [1] CRAN (R 4.2.1)
 RColorBrewer      1.1-3      2022-04-03 [1] CRAN (R 4.2.0)
 Rcpp              1.0.12     2024-01-09 [1] CRAN (R 4.2.3)
 readr           * 2.1.5      2024-01-10 [1] CRAN (R 4.2.3)
 related         * 1.0        2022-07-14 [1] local
 remotes           2.4.2.1    2023-07-18 [1] CRAN (R 4.2.3)
 reshape2          1.4.4      2020-04-09 [1] CRAN (R 4.2.1)
 rlang             1.1.3      2024-01-10 [1] CRAN (R 4.2.3)
 rmarkdown         2.25       2023-09-18 [1] CRAN (R 4.2.3)
 robustbase        0.99-1     2023-11-29 [1] CRAN (R 4.2.3)
 rpart             4.1.23     2023-12-05 [1] CRAN (R 4.2.3)
 rstudioapi        0.15.0     2023-07-07 [1] CRAN (R 4.2.3)
 scales            1.3.0      2023-11-28 [1] CRAN (R 4.2.3)
 seqinr            4.2-36     2023-12-08 [1] CRAN (R 4.2.3)
 sessioninfo       1.2.2      2021-12-06 [1] CRAN (R 4.2.3)
 shiny             1.8.0      2023-11-17 [1] CRAN (R 4.2.3)
 SNPRelate         1.36.1     2024-01-25 [1] Github (zhengxwen/SNPRelate@19af345)
 snpStats        * 1.48.0     2022-11-01 [1] Bioconductor
 StAMPP            1.6.3      2021-08-08 [1] CRAN (R 4.2.1)
 stringi           1.8.3      2023-12-11 [1] CRAN (R 4.2.3)
 stringr         * 1.5.1      2023-11-14 [1] CRAN (R 4.2.3)
 survival        * 3.5-7      2023-08-14 [1] CRAN (R 4.2.3)
 svglite           2.1.3      2023-12-08 [1] CRAN (R 4.2.3)
 systemfonts       1.0.5      2023-10-09 [1] CRAN (R 4.2.3)
 tensorA           0.36.2.1   2023-12-13 [1] CRAN (R 4.2.3)
 tibble          * 3.2.1      2023-03-20 [1] CRAN (R 4.2.3)
 tidyr           * 1.3.1      2024-01-24 [1] CRAN (R 4.2.1)
 tidyselect        1.2.0      2022-10-10 [1] CRAN (R 4.2.3)
 tidyverse       * 2.0.0      2023-02-22 [1] CRAN (R 4.2.3)
 timechange        0.3.0      2024-01-18 [1] CRAN (R 4.2.3)
 tzdb              0.4.0      2023-05-12 [1] CRAN (R 4.2.3)
 urlchecker        1.0.1      2021-11-30 [1] CRAN (R 4.2.3)
 usethis           2.2.2      2023-07-06 [1] CRAN (R 4.2.3)
 utf8              1.2.4      2023-10-22 [1] CRAN (R 4.2.3)
 vctrs             0.6.5      2023-12-01 [1] CRAN (R 4.2.3)
 vegan             2.6-4      2022-10-11 [1] CRAN (R 4.2.3)
 viridisLite       0.4.2      2023-05-02 [1] CRAN (R 4.2.3)
 withr             3.0.0      2024-01-16 [1] CRAN (R 4.2.3)
 xfun              0.41       2023-11-01 [1] CRAN (R 4.2.3)
 xml2              1.3.6      2023-12-04 [1] CRAN (R 4.2.3)
 xtable            1.8-4      2019-04-21 [1] CRAN (R 4.2.1)
 yaml              2.3.8      2023-12-11 [1] CRAN (R 4.2.3)
 zlibbioc          1.44.0     2022-11-01 [1] Bioconductor

 [1] C:/Users/dev093/AppData/Local/R/win-library/4.2
 [2] C:/Program Files/R/R-4.2.1/library

──────────────────────────────────────────────────────────────────────────────────────────────────</code></pre>
</div>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" role="list" style="display: none">
<div id="ref-devloo_bull" class="csl-entry" role="listitem">
Devloo-Delva, Floriaan, Christopher P. Burridge, Peter M. Kyne, Juerg M. Brunnschweiler, Demian D. Chapman, Patricia Charvet, Xiao Chen, et al. 2023. <span>“From Rivers to Ocean Basins: The Role of Ocean Barriers and Philopatry in the Genetic Structuring of a Cosmopolitan Coastal Predator.”</span> Journal Article. <em>Ecology and Evolution</em> 13 (2): e9837. https://doi.org/<a href="https://doi.org/10.1002/ece3.9837">https://doi.org/10.1002/ece3.9837</a>.
</div>
<div id="ref-devloo_slm" class="csl-entry" role="listitem">
Devloo-Delva, Floriaan, Thierry Gosselin, Paul A Butcher, Peter M Grewe, Charlie Huveneers, Robin B Thomson, Jonathan M Werry, and Pierre Feutry. 2023. <span>“An r-Based Tool for Identifying Sex-Linked Markers from Restriction Site-Associated DNA Sequencing with Applications to Elasmobranch Conservation.”</span> Journal Article. <em>Conservation Genetics Resources</em> 16: 11–16. <a href="https://doi.org/10.1007/s12686-023-01331-5">https://doi.org/10.1007/s12686-023-01331-5</a>.
</div>
<div id="ref-feutry_glyphis" class="csl-entry" role="listitem">
Feutry, Pierre, Floriaan Devloo-Delva, Adrien Tran Lu Y, Stefano Mona, Rasanthi M. Gunasekera, Grant Johnson, Richard D. Pillans, et al. 2020. <span>“One Panel to Rule Them All: DArTcap Genotyping for Population Structure, Historical Demography, and Kinship Analyses, and Its Application to a Threatened Shark.”</span> Journal Article. <em>Molecular Ecology Resources</em> 20 (6): 1470–85. <a href="https://doi.org/10.1111/1755-0998.13204">https://doi.org/10.1111/1755-0998.13204</a>.
</div>
<div id="ref-forester_gea" class="csl-entry" role="listitem">
Forester, Brenna R, Jesse R Lasky, Helene H Wagner, and Dean L Urban. 2018. <span>“Comparing Methods for Detecting Multilocus Adaptation with Multivariate Genotype–Environment Associations.”</span> Journal Article. <em>Molecular Ecology</em> 27 (9): 2215–33.
</div>
<div id="ref-Georges2023" class="csl-entry" role="listitem">
Georges, Arthur, Luis Mijangos, Andrzej Kilian, Hardip Patel, Mark Aitkens, and Bernd Gruber. 2023. <span>“Distances and Their Visualization in Studies of Spatial-Temporal Genetic Variation Using Single Nucleotide Polymorphisms (SNPs).”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2023.03.22.533737">https://doi.org/10.1101/2023.03.22.533737</a>.
</div>
<div id="ref-goudet_kiship" class="csl-entry" role="listitem">
Goudet, Jérôme, Tomas Kay, and Bruce S. Weir. 2018. <span>“How to Estimate Kinship.”</span> <em>Molecular Ecology</em> 27 (20): 4121–35. https://doi.org/<a href="https://doi.org/10.1111/mec.14833">https://doi.org/10.1111/mec.14833</a>.
</div>
<div id="ref-gruber_dartr" class="csl-entry" role="listitem">
Gruber, B., P. J. Unmack, O. F. Berry, and A. Georges. 2018. <span>“Dartr: An r Package to Facilitate Analysis of SNP Data Generated from Reduced Representation Genome Sequencing.”</span> Journal Article. <em>Molecular Ecology Resources</em> 18 (3): 691–99. <a href="https://doi.org/10.1111/1755-0998.12745">https://doi.org/10.1111/1755-0998.12745</a>.
</div>
<div id="ref-jost2008gst" class="csl-entry" role="listitem">
Jost, Lou. 2008. <span>“GST and Its Relatives Do Not Measure Differentiation.”</span> <em>Molecular Ecology</em> 17 (18): 4015–26. <a href="https://doi.org/10.1111/j.1365-294X.2008.03887.x">https://doi.org/10.1111/j.1365-294X.2008.03887.x</a>.
</div>
<div id="ref-mijangos_dartr2" class="csl-entry" role="listitem">
Mijangos, Jose Luis, Bernd Gruber, Oliver Berry, Carlo Pacioni, and Arthur Georges. 2022. <span>“dartR V2: An Accessible Genetic Analysis Platform for Conservation, Ecology and Agriculture.”</span> Journal Article. <em>Methods in Ecology and Evolution</em> 13 (10): 2150–58. https://doi.org/<a href="https://doi.org/10.1111/2041-210X.13918">https://doi.org/10.1111/2041-210X.13918</a>.
</div>
<div id="ref-nikolic_blue" class="csl-entry" role="listitem">
Nikolic, Natacha, Floriaan Devloo-Delva, Diane Bailleul, Ekaterina Noskova, Clément Rougeux, Chrystelle Delord, Philippe Borsa, et al. 2023. <span>“Stepping up to Genome Scan Allows Stock Differentiation in the Worldwide Distributed Blue Shark Prionace Glauca.”</span> Journal Article. <em>Molecular Ecology</em> 32 (5): 1000–1019. https://doi.org/<a href="https://doi.org/10.1111/mec.16822">https://doi.org/10.1111/mec.16822</a>.
</div>
<div id="ref-pew_related" class="csl-entry" role="listitem">
Pew, Jack, Paul H Muir, Jinliang Wang, and Timothy R Frasier. 2015. <span>“Related: An r Package for Analysing Pairwise Relatedness from Codominant Molecular Markers.”</span> Journal Article. <em>Molecular Ecology Resources</em> 15 (3): 557–61. https://doi.org/<a href="https://doi.org/10.1111/1755-0998.12323">https://doi.org/10.1111/1755-0998.12323</a>.
</div>
<div id="ref-robledo_slm" class="csl-entry" role="listitem">
Robledo-Ruiz, Diana A., Lana Austin, J. Nevil Amos, Jesús Castrejón-Figueroa, Daniel K. P. Harley, Michael J. L. Magrath, Paul Sunnucks, and Alexandra Pavlova. 2023. <span>“Easy-to-Use r Functions to Separate Reduced-Representation Genomic Datasets into Sex-Linked&nbsp;and Autosomal Loci, and Conduct Sex Assignment.”</span> Journal Article. <em>Molecular Ecology Resources</em> 00: 1–21. https://doi.org/<a href="https://doi.org/10.1111/1755-0998.13844">https://doi.org/10.1111/1755-0998.13844</a>.
</div>
<div id="ref-sherwin" class="csl-entry" role="listitem">
Sherwin, William B, Anne Chao, Lou Jost, and Peter E Smouse. 2017. <span>“Information Theory Broadens the Spectrum of Molecular Ecology and Evolution.”</span> Journal Article. <em>Trends in Ecology &amp; Evolution</em> 32 (12): 948–63. https://doi.org/<a href="https://doi.org/10.1016/j.tree.2017.09.012">https://doi.org/10.1016/j.tree.2017.09.012</a>.
</div>
<div id="ref-speed2015_relatedness" class="csl-entry" role="listitem">
Speed, Doug, and David J Balding. 2015. <span>“Relatedness in the Post-Genomic Era: Is It Still Useful?”</span> <em>Nature Reviews Genetics</em> 16 (1): 33–44. <a href="https://doi.org/10.1038/nrg3821">https://doi.org/10.1038/nrg3821</a>.
</div>
<div id="ref-sundqvist_divmigrate" class="csl-entry" role="listitem">
Sundqvist, Lisa, Kevin Keenan, Martin Zackrisson, Paulo Prodöhl, and David Kleinhans. 2016. <span>“Directional Genetic Differentiation and Relative Migration.”</span> <em>Ecology and Evolution</em> 6 (11): 3461–75. https://doi.org/<a href="https://doi.org/10.1002/ece3.2096">https://doi.org/10.1002/ece3.2096</a>.
</div>
<div id="ref-wang2022_kinship" class="csl-entry" role="listitem">
Wang, Jinliang. 2022. <span>“A Joint Likelihood Estimator of Relatedness and Allele Frequencies from a Small Sample of Individuals.”</span> <em>Methods in Ecology and Evolution</em> 13 (11): 2443–62. https://doi.org/<a href="https://doi.org/10.1111/2041-210X.13963">https://doi.org/10.1111/2041-210X.13963</a>.
</div>
<div id="ref-waples_hwe" class="csl-entry" role="listitem">
Waples, R. S. 2015. <span>“Testing for Hardy-Weinberg Proportions: Have We Lost the Plot?”</span> Journal Article. <em>Journal of Heredity</em> 106 (1): 1–19. <a href="https://doi.org/10.1093/jhered/esu062">https://doi.org/10.1093/jhered/esu062</a>.
</div>
<div id="ref-whitlock_outliers" class="csl-entry" role="listitem">
Whitlock, M. C., and K. E. Lotterhos. 2015. <span>“Reliable Detection of Loci Responsible for Local Adaptation: Inference of a Null Model Through Trimming the Distribution of Fst.”</span> Journal Article. <em>The American Naturalist</em> 186 (S1): S24–36. <a href="https://doi.org/10.1086/682949">https://doi.org/10.1086/682949</a>.
</div>
</div>
</section>

</main> <!-- /main -->
<script>

/* update total correct if #webex-total_correct exists */
update_total_correct = function() {
  console.log("webex: update total_correct");

  var t = document.getElementsByClassName("webex-total_correct");
  for (var i = 0; i < t.length; i++) {
    p = t[i].parentElement;
    var correct = p.getElementsByClassName("webex-correct").length;
    var solvemes = p.getElementsByClassName("webex-solveme").length;
    var radiogroups = p.getElementsByClassName("webex-radiogroup").length;
    var selects = p.getElementsByClassName("webex-select").length;

    t[i].innerHTML = correct + " of " + (solvemes + radiogroups + selects) + " correct";
  }
}

/* webex-solution button toggling function */
b_func = function() {
  console.log("webex: toggle hide");

  var cl = this.parentElement.classList;
  if (cl.contains('open')) {
    cl.remove("open");
  } else {
    cl.add("open");
  }
}

/* check answers */
check_func = function() {
  console.log("webex: check answers");

  var cl = this.parentElement.classList;
  if (cl.contains('unchecked')) {
    cl.remove("unchecked");
    this.innerHTML = "Hide Answers";
  } else {
    cl.add("unchecked");
    this.innerHTML = "Show Answers";
  }
}

/* function for checking solveme answers */
solveme_func = function(e) {
  console.log("webex: check solveme");

  var real_answers = JSON.parse(this.dataset.answer);
  var my_answer = this.value;
  var cl = this.classList;
  if (cl.contains("ignorecase")) {
    my_answer = my_answer.toLowerCase();
  }
  if (cl.contains("nospaces")) {
    my_answer = my_answer.replace(/ /g, "")
  }

  if (my_answer == "") {
    cl.remove("webex-correct");
    cl.remove("webex-incorrect");
  } else if (real_answers.includes(my_answer)) {
    cl.add("webex-correct");
    cl.remove("webex-incorrect");
  } else {
    cl.add("webex-incorrect");
    cl.remove("webex-correct");
  }

  // match numeric answers within a specified tolerance
  if(this.dataset.tol > 0){
    var tol = JSON.parse(this.dataset.tol);
    var matches = real_answers.map(x => Math.abs(x - my_answer) < tol)
    if (matches.reduce((a, b) => a + b, 0) > 0) {
      cl.add("webex-correct");
    } else {
      cl.remove("webex-correct");
    }
  }

  // added regex bit
  if (cl.contains("regex")){
    answer_regex = RegExp(real_answers.join("|"))
    if (answer_regex.test(my_answer)) {
      cl.add("webex-correct");
    }
  }

  update_total_correct();
}

/* function for checking select answers */
select_func = function(e) {
  console.log("webex: check select");

  var cl = this.classList

  /* add style */
  cl.remove("webex-incorrect");
  cl.remove("webex-correct");
  if (this.value == "answer") {
    cl.add("webex-correct");
  } else if (this.value != "blank") {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

/* function for checking radiogroups answers */
radiogroups_func = function(e) {
  console.log("webex: check radiogroups");

  var checked_button = document.querySelector('input[name=' + this.id + ']:checked');
  var cl = checked_button.parentElement.classList;
  var labels = checked_button.parentElement.parentElement.children;

  /* get rid of styles */
  for (i = 0; i < labels.length; i++) {
    labels[i].classList.remove("webex-incorrect");
    labels[i].classList.remove("webex-correct");
  }

  /* add style */
  if (checked_button.value == "answer") {
    cl.add("webex-correct");
  } else {
    cl.add("webex-incorrect");
  }

  update_total_correct();
}

window.onload = function() {
  console.log("webex onload");
  /* set up solution buttons */
  var buttons = document.getElementsByTagName("button");

  for (var i = 0; i < buttons.length; i++) {
    if (buttons[i].parentElement.classList.contains('webex-solution')) {
      buttons[i].onclick = b_func;
    }
  }

  var check_sections = document.getElementsByClassName("webex-check");
  console.log("check:", check_sections.length);
  for (var i = 0; i < check_sections.length; i++) {
    check_sections[i].classList.add("unchecked");

    let btn = document.createElement("button");
    btn.innerHTML = "Show Answers";
    btn.classList.add("webex-check-button");
    btn.onclick = check_func;
    check_sections[i].appendChild(btn);

    let spn = document.createElement("span");
    spn.classList.add("webex-total_correct");
    check_sections[i].appendChild(spn);
  }

  /* set up webex-solveme inputs */
  var solveme = document.getElementsByClassName("webex-solveme");

  for (var i = 0; i < solveme.length; i++) {
    /* make sure input boxes don't auto-anything */
    solveme[i].setAttribute("autocomplete","off");
    solveme[i].setAttribute("autocorrect", "off");
    solveme[i].setAttribute("autocapitalize", "off");
    solveme[i].setAttribute("spellcheck", "false");
    solveme[i].value = "";

    /* adjust answer for ignorecase or nospaces */
    var cl = solveme[i].classList;
    var real_answer = solveme[i].dataset.answer;
    if (cl.contains("ignorecase")) {
      real_answer = real_answer.toLowerCase();
    }
    if (cl.contains("nospaces")) {
      real_answer = real_answer.replace(/ /g, "");
    }
    solveme[i].dataset.answer = real_answer;

    /* attach checking function */
    solveme[i].onkeyup = solveme_func;
    solveme[i].onchange = solveme_func;

    solveme[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  /* set up radiogroups */
  var radiogroups = document.getElementsByClassName("webex-radiogroup");
  for (var i = 0; i < radiogroups.length; i++) {
    radiogroups[i].onchange = radiogroups_func;
  }

  /* set up selects */
  var selects = document.getElementsByClassName("webex-select");
  for (var i = 0; i < selects.length; i++) {
    selects[i].onchange = select_func;
    selects[i].insertAdjacentHTML("afterend", " <span class='webex-icon'></span>")
  }

  update_total_correct();
}

</script>
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
          // target, if specified
          link.setAttribute("target", "_blank");
          // default icon
          link.classList.add("external");
      }
    }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./DArT_dartR.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">DArT and dartR background</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./references.html" class="pagination-link">
        <span class="nav-page-text">References</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>